{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vemonet/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1/52] The orthologs of a gene that is expressed in the fruit fly’s brain and the UniProt annotations of these orthologs.\n",
      "0 / 1\n",
      "[2/52] Which are the genes in Primates orthologous to a gene that is expressed in the fruit fly’s brain and the UniProt annotations of the genes.\n",
      "0 / 2\n",
      "[3/52] Which are the genes in Human associated to a disease that are orthologous to a gene expressed in the rat’s brain?\n",
      "https://www.bgee.org/sparql/\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "SELECT ?gene ?OMA_LINK ?uniprot ?annotation_text WHERE {\n",
      "\tSERVICE <https://www.bgee.org/sparql/> {\n",
      "\t\tSELECT ?gene {\n",
      "\t\t\t?gene genex:isExpressedIn ?anat .\n",
      "\t\t\t?anat rdfs:label 'brain' .\n",
      "\t\t\t?gene orth:organism ?o .\n",
      "\t\t\t?o obo:RO_0002162 ?taxon .\n",
      "\t\t\t?taxon up:commonName 'rat' .\n",
      "\t\t}\n",
      "\t\tLIMIT 10\n",
      "\t}\n",
      "\tSERVICE <http://sparql.uniprot.org/sparql> {\n",
      "\t\t?taxon2 up:commonName 'Human' .\n",
      "\t}\n",
      "\tSERVICE <https://sparql.omabrowser.org/lode/sparql> {\n",
      "\t\t?protein2 a orth:Protein .\n",
      "\t\t?protein1 a orth:Protein .\n",
      "\t\t?cluster a orth:OrthologsCluster .\n",
      "\t\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t\t?node2 orth:hasHomologousMember* ?protein2 .\n",
      "\t\t?node1 orth:hasHomologousMember* ?protein1 .\n",
      "\t\t?protein1 sio:SIO_010079 ?gene. #encoded by\n",
      "\t\t?protein2 rdfs:seeAlso ?OMA_LINK .\n",
      "\t\t?protein2 orth:organism/obo:RO_0002162 ?taxon2 .\n",
      "\t\t?protein2 lscr:xrefUniprot ?uniprot .\n",
      "\t\tFILTER ( ?node1 != ?node2 )\n",
      "\t}\n",
      "\tSERVICE <http://sparql.uniprot.org/sparql> {\n",
      "\t\t?uniprot up:annotation ?annotation .\n",
      "\t\t?annotation rdfs:comment ?annotation_text .\n",
      "\t\t?annotation a up:Disease_Annotation .\n",
      "\t}\n",
      "}\n",
      "1 / 2\n",
      "[4/52] Which are the proteins associated with \"glioblastoma\" and the orthologs expressed in the rat’s brain?\n",
      "1 / 3\n",
      "[5/52] What are the Homo sapiens’ genes associated with cancer and their orthologs expressed in the Rattus norvegicus’s brain?\n",
      "1 / 4\n",
      "[6/52] Anatomic entities where the ins zebrafish gene is expressed and its gene GO annotations.\n",
      "https://www.bgee.org/sparql/\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT DISTINCT ?anat_label ?GO_CLASS ?GO_LABEL {\n",
      "\tSERVICE <https://www.bgee.org/sparql/> {\n",
      "\t\tSELECT ?geneEns ?anat_label {\n",
      "\t\t\t?geneB a orth:Gene .\n",
      "\t\t\t?geneB genex:isExpressedIn ?anat .\n",
      "\t\t\t?geneB rdfs:label ?anatLabel ;\n",
      "\t\t\t\tlscr:xrefEnsemblGene ?geneEns .\n",
      "\t\t\tFILTER (UCASE($anatLabel) = UCASE('ins'))\n",
      "\t\t\t?anat rdfs:label ?anat_label .\n",
      "\t\t\t?geneB orth:organism ?o .\n",
      "\t\t\t?o obo:RO_0002162 ?taxon2 .\n",
      "\t\t\t?taxon2 up:commonName 'zebrafish' .\n",
      "\t\t}\n",
      "\t}\n",
      "\tSERVICE <http://sparql.uniprot.org/sparql> {\n",
      "\t\t?uniprot rdfs:seeAlso / up:transcribedFrom ?geneEns .\n",
      "\t\t?uniprot a up:Protein .\n",
      "\t\t?uniprot up:classifiedWith ?GO_CLASS .\n",
      "\t\t?GO_CLASS rdfs:label ?GO_LABEL .\n",
      "\t}\n",
      "}\n",
      "LIMIT 10\n",
      "2 / 4\n",
      "[7/52] Genes expressed in the human's pancreas and their annotations in UniProt.\n",
      "2 / 5\n",
      "[8/52] Genes expressed in the human's brain during the infant stage and their UniProt disease annotations\n",
      "2 / 6\n",
      "[9/52] Proteins in OMA that are encoded by the INS gene without considering synonyms and their species identifier and evidence types from UniProt database.\n",
      "https://www.bgee.org/sparql/\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?proteinOMA ?species ?mnemonic ?evidenceType ?UniProt_URI WHERE {\n",
      "\tSERVICE <https://sparql.omabrowser.org/sparql/> {\n",
      "\t\t?proteinOMA a orth:Protein ;\n",
      "\t\t\torth:organism/obo:RO_0002162/up:scientificName ?species ;\n",
      "\t\t\trdfs:label 'INS' .\n",
      "\t\t?proteinOMA lscr:xrefUniprot ?UniProt_URI .\n",
      "\t}\n",
      "\tSERVICE <http://sparql.uniprot.org/sparql> {\n",
      "\t\t?UniProt_URI up:mnemonic ?mnemonic ;\n",
      "\t\t\tup:existence/rdfs:label ?evidenceType .\n",
      "\t}\n",
      "}\n",
      "3 / 6\n",
      "[10/52] Rabbit's proteins encoded by genes that are orthologous to Mouse's HBB-Y gene and their cross reference links to Uniprot\n",
      "https://www.bgee.org/sparql/\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?PROTEIN_1 ?PROTEIN_2 ?UNIPROT_XREF_1 ?UNIPROT_XREF_2 WHERE {\n",
      "\tSERVICE <http://sparql.uniprot.org/sparql> {\n",
      "\t\t?taxon_1 up:commonName 'Mouse' .\n",
      "\t\t?taxon_2 up:commonName 'Rabbit' .\n",
      "\t}\n",
      "\tSERVICE <https://sparql.omabrowser.org/sparql/> {\n",
      "\t\t?cluster a orth:OrthologsCluster .\n",
      "\t\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t\t?node2 orth:hasHomologousMember* ?PROTEIN_2 .\n",
      "\t\t?node1 orth:hasHomologousMember* ?PROTEIN_1 .\n",
      "\t\t?PROTEIN_1 a orth:Protein .\n",
      "\t\t?PROTEIN_1 orth:organism/obo:RO_0002162 ?taxon_1 ;\n",
      "\t\t\trdfs:label 'HBB-Y' ;\n",
      "\t\t\tlscr:xrefUniprot ?UNIPROT_XREF_1 .\n",
      "\t\t?PROTEIN_2 a orth:Protein .\n",
      "\t\t?PROTEIN_2 orth:organism/obo:RO_0002162 ?taxon_2 .\n",
      "\t\t?PROTEIN_2 lscr:xrefUniprot ?UNIPROT_XREF_2 .\n",
      "\t\tFILTER ( ?node1 != ?node2 )\n",
      "\t}\n",
      "}\n",
      "4 / 6\n",
      "[11/52] Rattus norvegicus' proteins encoded by genes that are paralogous to its TP53 gene and their Uniprot function annotations.\n",
      "4 / 7\n",
      "[12/52] Which are the mouse's proteins encoded by genes which are expressed in the liver and are orthologous to human's INS gene?\n",
      "https://www.bgee.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "SELECT ?name1 ?protein1 ?name2 ?protein2 ?OMA_link2 ?anatomicalEntity {\n",
      "\tSELECT DISTINCT * {\n",
      "\t\tSERVICE <https://www.bgee.org/sparql/> {\n",
      "\t\t\t?taxon up:commonName 'human' ;\n",
      "\t\t\t\tup:commonName ?name1 .\n",
      "\t\t\t?taxon2 up:commonName 'mouse' ;\n",
      "\t\t\t\tup:commonName ?name2 .\n",
      "\t\t}\n",
      "\t\tSERVICE <https://sparql.omabrowser.org/sparql/> {\n",
      "\t\t\t?cluster a orth:OrthologsCluster .\n",
      "\t\t\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t\t\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t\t\t?node2 orth:hasHomologousMember* ?protein2 .\n",
      "\t\t\t?node1 orth:hasHomologousMember* ?protein1 .\n",
      "\t\t\t?protein1 a orth:Protein .\n",
      "\t\t\t?protein1 rdfs:label 'INS' ;\n",
      "\t\t\t\torth:organism/obo:RO_0002162 ?taxon .\n",
      "\t\t\t?protein2 a orth:Protein ;\n",
      "\t\t\t\tsio:SIO_010079 ?gene ; #is encoded by\n",
      "\t\t\t\torth:organism/obo:RO_0002162 ?taxon2 .\n",
      "\t\t\t?gene lscr:xrefEnsemblGene ?geneEns .\n",
      "\t\t\t?protein2 rdfs:seeAlso ?OMA_link2 .\n",
      "\t\t\tFILTER ( ?node1 != ?node2 )\n",
      "\t\t}\n",
      "\t\tSERVICE <https://www.bgee.org/sparql/> {\n",
      "\t\t\t?geneB a orth:Gene .\n",
      "\t\t\t\t?geneB genex:isExpressedIn ?cond .\n",
      "\t\t\t\t?cond genex:hasAnatomicalEntity ?anat .\n",
      "\t\t\t\t?geneB lscr:xrefEnsemblGene ?geneEns .\n",
      "\t\t\t?anat rdfs:label 'liver' ;\n",
      "\t\t\t\trdfs:label ?anatomicalEntity .\n",
      "\t\t\t?geneB orth:organism ?o .\n",
      "\t\t\t?o obo:RO_0002162 ?taxon2 .\n",
      "\t\t}\n",
      "\t}\n",
      "\tLIMIT 10\n",
      "}\n",
      "LIMIT 10\n",
      "5 / 7\n",
      "[13/52] Which are the orthologs of a gene that is expressed in the fruit fly’s brain?\n",
      "https://www.bgee.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?id ?OMA_LINK WHERE {\n",
      "\tSELECT * {\n",
      "\t\tSERVICE <https://www.bgee.org/sparql/> {\n",
      "\t\t\tSELECT DISTINCT ?gene ?id {\n",
      "\t\t\t\t?gene a orth:Gene .\n",
      "\t\t\t\t?gene genex:isExpressedIn ?anat .\n",
      "\t\t\t\t?anat rdfs:label 'brain' .\n",
      "\t\t\t\t?gene orth:organism ?o .\n",
      "\t\t\t\t?o obo:RO_0002162 ?taxon .\n",
      "\t\t\t\t?gene dcterms:identifier ?id .\n",
      "\t\t\t\t?taxon up:commonName 'fruit fly' .\n",
      "\t\t\t}\n",
      "\t\t\tLIMIT 100\n",
      "\t\t}\n",
      "\t\tSERVICE <https://sparql.omabrowser.org/lode/sparql> {\n",
      "\t\t\t?cluster a orth:OrthologsCluster .\n",
      "\t\t\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t\t\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t\t\t?node2 orth:hasHomologousMember* ?protein2 .\n",
      "\t\t\t?node1 orth:hasHomologousMember* ?protein1 .\n",
      "\t\t\t?protein1 dcterms:identifier ?id .\n",
      "\t\t\t?protein2 rdfs:seeAlso ?OMA_LINK .\n",
      "\t\t\tFILTER ( ?node1 != ?node2 )\n",
      "\t\t}\n",
      "\t}\n",
      "}\n",
      "LIMIT 10\n",
      "6 / 7\n",
      "[14/52] Which are the genes in Primates orthologous to a gene that is expressed in the fruit fly’s eye?\n",
      "https://www.bgee.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?orthologous_to_gene_id ?orthologous_to_gene ?hasOrtholog WHERE {\n",
      "\tSELECT * {\n",
      "\t\tSERVICE <https://www.bgee.org/sparql/> {\n",
      "\t\t\tSELECT DISTINCT ?gene ?orthologous_to_gene_id ?orthologous_to_gene {\n",
      "\t\t\t\t?gene a orth:Gene .\n",
      "\t\t\t\t?gene genex:isExpressedIn ?anat .\n",
      "\t\t\t\t?anat rdfs:label 'eye' .\n",
      "\t\t\t\t?gene orth:organism ?o .\n",
      "\t\t\t\t?o obo:RO_0002162 ?taxon .\n",
      "\t\t\t\t?gene dcterms:identifier ?orthologous_to_gene_id .\n",
      "\t\t\t\t?gene rdfs:seeAlso ?orthologous_to_gene .\n",
      "\t\t\t\t?taxon up:commonName 'fruit fly' .\n",
      "\t\t\t} LIMIT 100\n",
      "\t\t}\n",
      "\t\tSERVICE <https://sparql.omabrowser.org/sparql/> {\n",
      "\t\t\t?protein2 a orth:Protein .\n",
      "\t\t\t?protein1 a orth:Protein .\n",
      "\t\t\t?clusterPrimates a orth:OrthologsCluster .\n",
      "\t\t\t?cluster a orth:OrthologsCluster .\n",
      "\t\t\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t\t\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t\t\t?node2 orth:hasHomologousMember* ?clusterPrimates .\n",
      "\t\t\t?clusterPrimates orth:hasHomologousMember* ?protein2 .\n",
      "\t\t\t?node1 orth:hasHomologousMember* ?protein1 .\n",
      "\t\t\t?protein1 dcterms:identifier ?orthologous_to_gene_id .\n",
      "\t\t\t?protein2 rdfs:seeAlso ?hasOrtholog .\n",
      "\t\t\t?clusterPrimates orth:hasTaxonomicRange ?taxRange .\n",
      "\t\t\t?taxRange orth:taxRange ?taxLevel .\n",
      "\t\t\tVALUES ?taxLevel { 'Primates' }\n",
      "\t\t\tFILTER ( ?node1 != ?node2 )\n",
      "\t\t}\n",
      "\t}\n",
      "}\n",
      "LIMIT 10\n",
      "7 / 7\n",
      "[15/52] Scientific name of species in bgee with their uniprot taxon\n",
      "https://www.bgee.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT * {\n",
      "\t?species a up:Taxon .\n",
      "\t?species up:scientificName ?name .\n",
      "\t?species up:rank up:Species .\n",
      "}\n",
      "8 / 7\n",
      "[16/52] Genes with the following term in the description: lung\n",
      "https://www.bgee.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT ?geneName ?geneId {\n",
      "\t?gene a orth:Gene .\n",
      "\t?gene rdfs:label ?geneName .\n",
      "\t?gene dcterms:identifier ?geneId .\n",
      "\t?gene dcterms:description ?desc .\n",
      "\tFILTER CONTAINS ( ?desc, 'lung' )\n",
      "}\n",
      "9 / 7\n",
      "[17/52] Which species the Mt-co1 gene is present (without considering synonyms)?\n",
      "https://www.bgee.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT ?name {\n",
      "\t?gene a orth:Gene .\n",
      "\t?gene rdfs:label ?geneName .\n",
      "\t?gene orth:organism ?organism . #orth v2\n",
      "\t?organism obo:RO_0002162 ?taxon . #label: in taxon .\n",
      "\t?taxon up:scientificName ?name .\n",
      "\tFILTER ( UCASE(?geneName) = UCASE('Mt-co1') )\n",
      "}\n",
      "10 / 7\n",
      "[18/52] List the labels and identifiers of Fruit fly genes in bgee\n",
      "https://www.bgee.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT distinct ?gene_page ?geneName {\n",
      "\t?gene a orth:Gene .\n",
      "\t?gene rdfs:seeAlso ?gene_page .\n",
      "\t?gene rdfs:label ?geneName .\n",
      "\t?gene orth:organism ?taxon .\n",
      "\t?taxon obo:RO_0002162 ?species . #in taxon property .\n",
      "\t?species up:commonName ?commonName .\n",
      "\tFILTER ( lcase(str(?commonName)) = lcase(\"Fruit fly\") ) .\n",
      "}\n",
      "11 / 7\n",
      "[19/52] Human anatomic entities at young adult developmental stage\n",
      "https://www.bgee.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?anatName {\n",
      "\t?cond genex:hasAnatomicalEntity ?anatEntity .\n",
      "\t?anatEntity rdfs:label ?anatName .\n",
      "\t?cond genex:hasDevelopmentalStage ?stage .\n",
      "\t?stage rdfs:label ?stageName .\n",
      "\t?cond obo:RO_0002162 ?taxon .\n",
      "\t?taxon up:commonName ?commonName .\n",
      "\tFILTER ( lcase(str(?commonName)) = lcase(\"Human\") ).\n",
      "\tFILTER ( CONTAINS(?stageName, \"young adult\") )\n",
      "}\n",
      "12 / 7\n",
      "[20/52] Anatomic entities where the apoc1 gene is expressed\n",
      "https://www.bgee.org/sparql\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT ?anatEntity ?anatName {\n",
      "\t?seq a orth:Gene .\n",
      "\t?seq genex:isExpressedIn ?cond.\n",
      "\t?seq rdfs:label 'apoc1' .\n",
      "\t?cond genex:hasAnatomicalEntity ?anatEntity .\n",
      "\t?anatEntity rdfs:label ?anatName . \n",
      "} LIMIT 10\n",
      "13 / 7\n",
      "[21/52] Anatomic entities in rat where the apoc1 gene is expressed\n",
      "https://www.bgee.org/sparql\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?anatEntity ?anatName {\n",
      "\t?seq a orth:Gene .\n",
      "\t?seq genex:isExpressedIn ?cond .\n",
      "\t?seq rdfs:label ?geneName .\n",
      "\t?expr genex:hasExpressionCondition ?cond .\n",
      "\t?cond genex:hasAnatomicalEntity ?anatEntity .\n",
      "\t?anatEntity rdfs:label ?anatName .\n",
      "\t?cond obo:RO_0002162 <http://purl.uniprot.org/taxonomy/10116> . \n",
      "\tFILTER (LCASE(?geneName) = LCASE('apoc1') )\n",
      "}\n",
      "14 / 7\n",
      "[22/52] Anatomic entities at the Mouse’s adult stage where the apoc1 gene is expressed\n",
      "https://www.bgee.org/sparql\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?anatEntity ?anatName {\n",
      "\t?seq a orth:Gene .\n",
      "\t?seq genex:isExpressedIn ?cond .\n",
      "\t?seq rdfs:label ?geneName .\n",
      "\t?cond genex:hasAnatomicalEntity ?anatEntity .\n",
      "\t?anatEntity rdfs:label ?anatName .\n",
      "\t?cond genex:hasDevelopmentalStage ?stage .\n",
      "\t?stage rdfs:label\n",
      "\t?stageName .\n",
      "\t?cond obo:RO_0002162 ?taxon . #in taxon property .\n",
      "\t?taxon up:commonName ?commonName .\n",
      "\tFILTER ( lcase(str(?commonName)) = lcase(\"Mouse\") ) .\n",
      "\tFILTER ( CONTAINS(?stageName,'adult') && LCASE(?geneName) = LCASE('apoc1') )\n",
      "}\n",
      "15 / 7\n",
      "[23/52] Which species are available on OMA database and their scientific name?\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT * WHERE {\n",
      "\t?s a up:Taxon .\n",
      "\t?s up:scientificName ?v .\n",
      "\t?s up:rank up:Species .\n",
      "} \n",
      "16 / 7\n",
      "[24/52] Find Rattus norvegicus proteins present in OMA database\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?protein ?OMA_link WHERE {\n",
      "\t?protein a orth:Protein .\n",
      "\t?protein orth:organism ?organism .\n",
      "\t?inTaxon rdfs:label 'in taxon'@en .\n",
      "\t?organism ?inTaxon ?taxon .\n",
      "\t?taxon up:scientificName 'Rattus norvegicus' .\n",
      "\t?protein rdfs:seeAlso ?OMA_link .\n",
      "}\n",
      "17 / 7\n",
      "[25/52] How many proteins (and their scientific name) per species are present in OMA database?\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?taxon ?target (COUNT(?protein) as ?Protein_count) WHERE {\n",
      "\t?protein a orth:Protein .\n",
      "\t?protein orth:organism ?organism .\n",
      "\t?inTaxon rdfs:label 'in taxon'@en .\n",
      "\t?organism ?inTaxon ?taxon .\n",
      "\t?taxon up:scientificName ?target .\n",
      "}\n",
      "GROUP BY ?taxon ?target\n",
      "18 / 7\n",
      "[26/52] All proteins that are orthologous to ENSLACG00000002497 Ensembl gene (identifier) \n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?OMA_LINK WHERE {\n",
      "\t#The three that contains Orthologs. The leafs are proteins .\n",
      "\t#This graph pattern defines the relationship protein1 is Orthologs to protein2\n",
      "\t?cluster a orth:OrthologsCluster .\n",
      "\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t?node2 orth:hasHomologousMember* ?protein2 .\n",
      "\t?node1 orth:hasHomologousMember* ?protein1 .\n",
      "\t########\n",
      "\t#Specify the protein to look for its orthologs\n",
      "\t?protein1 sio:SIO_010079 ?gene .\n",
      "\t?gene lscr:xrefEnsemblGene ensembl:ENSLACG00000002497 .\n",
      "\t########\n",
      "\t#The OMA link to the second protein\n",
      "\t?protein2 rdfs:seeAlso ?OMA_LINK .\n",
      "\t########\n",
      "\tFILTER(?node1 != ?node2)\n",
      "}\n",
      "19 / 7\n",
      "[27/52] Proteins that are paralogous to ENSG00000244734 Ensembl gene (identifier)\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?OMA_LINK WHERE {\n",
      "\t#The three that contains paralogs. The leafs are proteins .\n",
      "\t#This graph pattern defines the relationship protein1 is paralogous to protein2\n",
      "\t?cluster a orth:ParalogsCluster .\n",
      "\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t?node2 orth:hasHomologousMember* ?protein2 .\n",
      "\t?node1 orth:hasHomologousMember* ?protein1 .\n",
      "\t########\n",
      "\t#Specify the protein to look for its paralogs\n",
      "\t?protein1 sio:SIO_010079 ?gene .\n",
      "\t?gene lscr:xrefEnsemblGene ensembl:ENSG00000244734 .\n",
      "\t########\n",
      "\t#The OMA link to the second protein\n",
      "\t?protein2 rdfs:seeAlso ?OMA_LINK .\n",
      "\t########\n",
      "\tFILTER(?node1 != ?node2)\n",
      "}\n",
      "20 / 7\n",
      "[28/52] Proteins that are paralogous to HUMAN04027 OMA protein (identifier) and their cross reference links to OMA and Uniprot.\n",
      "20 / 8\n",
      "[29/52] Proteins that are orthologous to HUMAN22168 OMA protein (identifier) and their cross reference links to OMA and Uniprot.\n",
      "20 / 9\n",
      "[30/52] All proteins per species that are orthologous to Oryctolagus cuniculus APOCI related protein and their cross reference links to OMA and Uniprot including the corresponding Ensembl gene identifier. \n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX dct: <http://purl.org/dc/terms/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "SELECT ?protein1 ?protein2 ?geneName2 ?species2 ?OMA_LINK2 ?Prot2_uniprot ?prot2_ensemblGeneId WHERE {\n",
      "\t?cluster a orth:OrthologsCluster .\n",
      "\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t?node2 orth:hasHomologousMember* ?protein2 .\n",
      "\t?node1 orth:hasHomologousMember* ?protein1 .\n",
      "\t?protein1 a orth:Protein ;\n",
      "\t\torth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus' ;\n",
      "\t\trdfs:label 'APOCI' .\n",
      "\t?protein2 a orth:Protein ;\n",
      "\t\trdfs:seeAlso ?OMA_LINK2 ;\n",
      "\t\tlscr:xrefUniprot ?Prot2_uniprot ;\n",
      "\t\tsio:SIO_010079/lscr:xrefEnsemblGene/dct:identifier ?prot2_ensemblGeneId ;\n",
      "\t\trdfs:label ?geneName2 ;\n",
      "\t\torth:organism/obo:RO_0002162/up:scientificName ?species2 .\n",
      "\tFILTER(?node1 != ?node2)\n",
      "}\n",
      "21 / 9\n",
      "[31/52] Oryctolagus cuniculus proteins (PROT1) encoded by genes that are orthologous to Mus musculus (PROT2) HBB-Y gene and their cross reference links to Uniprot\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX dct: <http://purl.org/dc/terms/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "SELECT DISTINCT ?PROT1 ?PROT2 ?PROT1_UNIPROT_XREF ?PROT2_UNIPROT_XREF WHERE {\n",
      "\t?cluster a orth:OrthologsCluster .\n",
      "\t?cluster orth:hasHomologousMember ?node1 .\n",
      "\t?cluster orth:hasHomologousMember ?node2 .\n",
      "\t?node2 orth:hasHomologousMember* ?PROT2 .\n",
      "\t?node1 orth:hasHomologousMember* ?PROT1 .\n",
      "\t?PROT1 a orth:Protein .\n",
      "\t?PROT1 orth:organism/obo:RO_0002162/up:scientificName 'Mus musculus' ;\n",
      "\t\trdfs:label 'HBB-Y' ;\n",
      "\t\tlscr:xrefUniprot ?PROT1_UNIPROT_XREF .\n",
      "\t?PROT2 a orth:Protein .\n",
      "\t?PROT2 orth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus' .\n",
      "\t?PROT2 lscr:xrefUniprot ?PROT2_UNIPROT_XREF .\n",
      "\tFILTER( ?node1 != ?node2 )\n",
      "}\n",
      "22 / 9\n",
      "[32/52] The percentage of proteins in Drosophila melanogaster that has at least one paralogous sequence (protein)\n",
      "22 / 10\n",
      "[33/52] Retrieve from Ensembl dataset at the EBI RDF platform the pairwise orthologs of the gene with the P68871 UniProt accession number - only works if you disable CORS in your browser\n",
      "22 / 11\n",
      "[34/52] Retrieve from Ensembl dataset at the EBI RDF platform the pairwise orthologs of the gene with the ENSG00000139618 Ensembl acession number - only works if you disable CORS in your browser\n",
      "22 / 12\n",
      "[35/52] Retrieve pairwise orthologs which are derived from the HOGs in the OMA database of gene with the P68871 UniProt accession number\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "SELECT DISTINCT ?protein1 ?protein2 {\n",
      "\tVALUES (?protein1){\n",
      "\t\t(<http://purl.uniprot.org/uniprot/P68871>)\n",
      "\t}\n",
      "\t?cluster a orth:OrthologsCluster.\n",
      "\t?cluster orth:hasHomologousMember ?node1.\n",
      "\t?cluster orth:hasHomologousMember ?node2.\n",
      "\t?node1 orth:hasHomologousMember* ?protein_OMA_1.\n",
      "\t?node2 orth:hasHomologousMember* ?protein_OMA_2.\n",
      "\t?protein_OMA_1 lscr:xrefUniprot ?protein1.\n",
      "\t?protein_OMA_2 lscr:xrefUniprot ?protein2.\n",
      "\tFILTER(?node1 != ?node2)\n",
      "}\n",
      "23 / 12\n",
      "[36/52] Retrieve pairwise orthologs which are derived from the HOGs in the MBGD database of gene with the P68871 UniProt accession number - only works if you disable CORS in your browser\n",
      "http://sparql.nibb.ac.jp/sparql\n",
      "PREFIX mbgdr: <http://mbgd.genome.ad.jp/rdf/resource/>\n",
      "PREFIX mbgd: <http://purl.jp/bio/11/mbgd#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT ?protein1 ?protein2 WHERE {\n",
      "\tVALUES (?protein1){\n",
      "\t\t(<http://purl.uniprot.org/uniprot/P68871>)\n",
      "\t}\n",
      "\t?cluster a orth:OrthologsCluster.\n",
      "\t?cluster orth:hasHomologous ?node1.\n",
      "\t?cluster orth:hasHomologous ?node2.\n",
      "\t?node1 orth:hasHomologous* ?gene1.\n",
      "\t?node2 orth:hasHomologous* ?gene2.\n",
      "\t?gene1 mbgd:uniprot ?protein1.\n",
      "\t?gene2 mbgd:uniprot ?protein2.\n",
      "\tFILTER(?node1 != ?node2)\n",
      "}\n",
      "24 / 12\n",
      "[37/52] Retrieve homologous groups (i.e. clusters) which are derived from the HOGs in the OMA database that contain the gene with the P68871 UniProt accession number\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?cluster ?protein2_OMA_URI ?protein2_uniprot_URI ?tax_name {\n",
      "\tVALUES (?protein1_uniprot_URI){\n",
      "\t\t(<http://purl.uniprot.org/uniprot/P68871>)\n",
      "\t}\n",
      "\tVALUES (?tax_name){\n",
      "\t\t(\"Primates\")\n",
      "\t}\n",
      "\t?cluster a orth:OrthologsCluster.\n",
      "\t?cluster orth:hasHomologousMember* ?protein_OMA_1.\n",
      "\t?cluster orth:hasHomologousMember* ?protein2_OMA_URI.\n",
      "\t?protein_OMA_1 a orth:Protein.\n",
      "\t?protein2_OMA_URI a orth:Protein.\n",
      "\t?protein_OMA_1 lscr:xrefUniprot ?protein1_uniprot_URI.\n",
      "\tOPTIONAL {\n",
      "\t\t?protein2_OMA_URI lscr:xrefUniprot ?protein2_uniprot_URI.\n",
      "\t}\n",
      "\t?cluster orth:hasTaxonomicRange ?tax.\n",
      "\t?tax orth:taxRange ?tax_name.\n",
      "}\n",
      "25 / 12\n",
      "[38/52] Retrieve flat homologous groups (i.e. clusters) which are derived from the HOGs in the OrthoDB that contain the gene with the P68871 UniProt accession number\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vemonet/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/SPARQLWrapper/Wrapper.py:1215: RuntimeWarning: unknown response content type 'text/html' returning raw response...\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25 / 13\n",
      "[39/52] Retrieve hierarchical orthologous groups from the OMA database that contain the gene with the P68871 UniProt accession number\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "SELECT DISTINCT ?root_hog ?species_name ?protein1_uniprot (?protein1 as ?protein1_OMA) ?taxLevel {\n",
      "\tVALUES ?protein2_uniprot {\n",
      "\t\t<http://purl.uniprot.org/uniprot/P68871>\n",
      "\t}\n",
      "\t?root_hog obo:CDAO_0000148 ?hog_cluster. #has_Root\n",
      "\t?hog_cluster orth:hasHomologousMember* ?node1.\n",
      "\t?node1 a orth:OrthologsCluster.\n",
      "\t?node1 orth:hasTaxonomicRange ?level.\n",
      "\t?level orth:taxRange ?taxLevel.\n",
      "\t?node1 orth:hasHomologousMember* ?protein1.\n",
      "\t?hog_cluster orth:hasHomologousMember* ?protein2.\n",
      "\t?protein1 a orth:Protein.\n",
      "\t?protein1 orth:organism ?organism.\n",
      "\t?organism obo:RO_0002162 ?taxon.\n",
      "\t?taxon up:scientificName ?species_name.\n",
      "\tOPTIONAL {\n",
      "\t\t?protein1 lscr:xrefUniprot ?protein1_uniprot\n",
      "\t}\n",
      "\t?protein2 a orth:Protein.\n",
      "\t?protein2 lscr:xrefUniprot ?protein2_uniprot.\n",
      "} ORDER BY ?taxLevel\n",
      "26 / 13\n",
      "[40/52] Retrieve hierarchical orthologous groups from the OrthoDB that contain the gene with the P68871 UniProt accession number - only works if you disable CORS in your browser\n",
      "https://sparql.orthodb.org/sparql\n",
      "PREFIX orthodb: <http://purl.orthodb.org/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT DISTINCT ?hog_root ?species_name ?protein1_uniprot ?gene1 ?taxLevel_uniprot ?taxLevel WHERE {\n",
      "\tVALUES ?protein2_uniprot {\n",
      "\t\t<http://purl.uniprot.org/uniprot/P68871>\n",
      "\t}\n",
      "\t?gene2 a orthodb:Gene.\n",
      "\t?gene2 orthodb:memberOf ?groups.\n",
      "\t?gene2 orthodb:memberOf ?hog_root.\n",
      "\tFILTER NOT EXISTS {\n",
      "\t\t?hog_root orthodb:ancestralOG ?ancestor.\n",
      "\t}\n",
      "\t?groups orthodb:ancestralOG* ?hog_root.\n",
      "\t?gene1 a orthodb:Gene.\n",
      "\t?gene1 orthodb:memberOf ?groups.\n",
      "\t?gene1 up:organism ?organism.\n",
      "\t?organism a ?taxon.\n",
      "\t?taxon up:scientificName ?species_name.\n",
      "\t?groups orthodb:ogBuiltAt ?taxLevel_uniprot.\n",
      "\t?taxLevel_uniprot up:scientificName ?taxLevel.\n",
      "\t?gene2 orthodb:xref ?xref2.\n",
      "\t?xref2 orthodb:xrefResource ?protein2_uniprot.\n",
      "\t?protein2_uniprot a orthodb:Uniprot.\n",
      "\t?gene1 orthodb:xref ?xref.\n",
      "\t?xref orthodb:xrefResource ?protein1_uniprot.\n",
      "\t?protein1_uniprot a orthodb:Uniprot.\n",
      "} ORDER BY ?hog_root ?taxLevel\n",
      "27 / 13\n",
      "[41/52] Retrieve pairwise orthologous genes of the Cyanobacterium-aponinum psb27- gene that are found in the MBGD database but are not present in OMA\n",
      "https://sparql.omabrowser.org/sparql/\n",
      "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX mbgd: <http://purl.jp/bio/11/mbgd#>\n",
      "SELECT ?protein2 ?species WHERE {\n",
      "\tSERVICE<http://sparql.nibb.ac.jp/sparql> {\n",
      "\t\tSELECT ?protein2 ?species where {\n",
      "\t\t\t?cluster_mbgd a orth:OrthologsCluster.\n",
      "\t\t\t?cluster_mbgd orth:hasHomologous ?node1_mbgd.\n",
      "\t\t\t?cluster_mbgd orth:hasHomologous ?node2_mbgd.\n",
      "\t\t\t?node1_mbgd orth:hasHomologous* ?gene1.\n",
      "\t\t\t?node2_mbgd orth:hasHomologous* ?gene2.\n",
      "\t\t\t?gene1 mbgd:uniprot <http://purl.uniprot.org/uniprot/K9Z723>.\n",
      "\t\t\t?gene2 mbgd:uniprot ?protein2.\n",
      "\t\t\t?gene2 mbgd:organism ?taxon.\n",
      "\t\t\tOPTIONAL {\n",
      "\t\t\t\t?taxon mbgd:species ?species.\n",
      "\t\t\t}\n",
      "\t\t\tFILTER (?node1_mbgd != ?node2_mbgd)\n",
      "\t\t}\n",
      "\t}\n",
      "\tFILTER NOT EXISTS { # keep only those that do not exist in OMA\n",
      "\t\t?cluster a orth:OrthologsCluster.\n",
      "\t\t?cluster orth:hasHomologousMember ?node1.\n",
      "\t\t?cluster orth:hasHomologousMember ?node2.\n",
      "\t\t?node1 orth:hasHomologousMember* ?protein_OMA_1.\n",
      "\t\t?node2 orth:hasHomologousMember* ?protein_OMA_2.\n",
      "\t\t?protein_OMA_1 lscr:xrefUniprot <http://purl.uniprot.org/uniprot/K9Z723>.\n",
      "\t\t?protein_OMA_2 lscr:xrefUniprot ?protein2.\n",
      "\t\tFILTER (?node1 != ?node2)\n",
      "\t}\n",
      "}\n",
      "28 / 13\n",
      "[42/52] Select the UniProt protein with the mnemonic 'A4_HUMAN'\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?protein WHERE\n",
      "{\n",
      "\t?protein a up:Protein .\n",
      "\t?protein up:mnemonic 'A4_HUMAN'\n",
      "}\n",
      "29 / 13\n",
      "[43/52] Select all UniProt proteins that were integrated on the 2010-11-30\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?protein WHERE {\n",
      "\t?protein a up:Protein .\n",
      "\t?protein up:created '2010-11-30'^^xsd:date\n",
      "}\n",
      "LIMIT 10\n",
      "30 / 13\n",
      "[44/52] Select all taxa from the UniProt taxonomy\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT * WHERE {\n",
      "\t?target a up:Taxon .\n",
      "}\n",
      "LIMIT 10\n",
      "31 / 13\n",
      "[45/52] Select all bacterial taxa, and their up:scientificName from the UniProt taxonomy\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?taxon ?target WHERE {\n",
      "\t?taxon a up:Taxon .\n",
      "\t?taxon up:scientificName ?target .\n",
      "\t?taxon rdfs:subClassOf taxon:2 .\n",
      "}\n",
      "LIMIT 10\n",
      "32 / 13\n",
      "[46/52] which are the Ensembl transcript and genes for the A4_HUMAN protein \n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT * WHERE {\n",
      "\t?protein a up:Protein .\n",
      "\t?protein up:mnemonic 'A4_HUMAN' .\n",
      "\t?protein rdfs:seeAlso ?ensembl_Transcript .\n",
      "\t?ensembl_Transcript a up:Transcript_Resource .\n",
      "\t?ensembl_Transcript up:transcribedFrom ?ensembl_Gene .\n",
      "\t?ensembl_Transcript up:database <http://purl.uniprot.org/database/Ensembl> .\n",
      "}\n",
      "33 / 13\n",
      "[47/52] Select all human UniProt entries with a sequence variant that leads to a loss of function\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?protein ?annotation WHERE {\n",
      "\t?protein a up:Protein .\n",
      "\t?protein up:organism taxon:9606 .\n",
      "\t?protein up:annotation ?annotation .\n",
      "\t?annotation a up:Natural_Variant_Annotation .\n",
      "\t?annotation rdfs:comment ?text .\n",
      "\tFILTER ( CONTAINS(?text,'loss of function') )\n",
      "}\n",
      "34 / 13\n",
      "[48/52] Select the preferred gene name and disease annotation of all human UniProt proteins that are known to be involved in a disease\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?name ?diseaseAnnotation WHERE {\n",
      "\t?protein a up:Protein .\n",
      "\t?protein up:organism taxon:9606 .\n",
      "\t?protein up:encodedBy ?gene .\n",
      "\t?gene skos:prefLabel ?name .\n",
      "\t?protein up:annotation ?annotation .\n",
      "\t?annotation a up:Disease_Annotation .\n",
      "\t?annotation rdfs:comment ?diseaseAnnotation .\n",
      "}\n",
      "35 / 13\n",
      "[49/52] Select all UniProt proteins with annotated transmembrane regions and the regions' begin and end coordinates on the canonical sequence\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX faldo: <http://biohackathon.org/resource/faldo#>\n",
      "SELECT ?protein ?begin ?end WHERE {\n",
      "\t?protein a up:Protein .\n",
      "\t?protein up:annotation ?annotation .\n",
      "\t?annotation a up:Transmembrane_Annotation .\n",
      "\t?annotation up:range ?range .\n",
      "\t?range faldo:begin/faldo:position ?begin .\n",
      "\t?range faldo:end/faldo:position ?end .\n",
      "} \n",
      "LIMIT 10\n",
      "36 / 13\n",
      "[50/52] Select the number of proteins for each of the subclasses of ec:1.1.1.- EC (Enzyme Commission) class\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX ec: <http://purl.uniprot.org/enzyme/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?enzyme (COUNT(?protein) as ?number) WHERE {\n",
      "\t?protein up:enzyme ?enzyme .\n",
      "\t?enzyme rdfs:subClassOf ec:1.1.1.- .\n",
      "}\n",
      "GROUP BY ?enzyme ORDER BY ?enzyme\n",
      "37 / 13\n",
      "[51/52] Select all E-Coli K12 (including strains) UniProt proteins and their amino acid sequence\n",
      "https://sparql.uniprot.org/sparql\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?protein ?aa WHERE {\n",
      "\t?protein a up:Protein .\n",
      "\t?protein up:organism ?organism .\n",
      "\t{\n",
      "\t\t?protein up:organism taxon:8333 .\n",
      "\t}\n",
      "\tUNION\n",
      "\t{\n",
      "\t\t?protein up:organism ?organism .\n",
      "\t\t?organism rdfs:subClassOf taxon:8333 .\n",
      "\t}\n",
      "\t?protein up:sequence ?s .\n",
      "\t?s rdf:value ?aa\n",
      "}\n",
      "38 / 13\n",
      "[\n",
      "  {\n",
      "    \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nSELECT ?gene ?OMA_LINK ?uniprot ?annotation_text WHERE {\\n\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\tSELECT ?gene {\\n\\t\\t\\t?gene genex:isExpressedIn ?anat .\\n\\t\\t\\t?anat rdfs:label 'brain' .\\n\\t\\t\\t?gene orth:organism ?o .\\n\\t\\t\\t?o obo:RO_0002162 ?taxon .\\n\\t\\t\\t?taxon up:commonName 'rat' .\\n\\t\\t}\\n\\t\\tLIMIT 10\\n\\t}\\n\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t?taxon2 up:commonName 'Human' .\\n\\t}\\n\\tSERVICE <https://sparql.omabrowser.org/lode/sparql> {\\n\\t\\t?protein2 a orth:Protein .\\n\\t\\t?protein1 a orth:Protein .\\n\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t?node2 orth:hasHomologousMember* ?protein2 .\\n\\t\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t\\t?protein1 sio:SIO_010079 ?gene. #encoded by\\n\\t\\t?protein2 rdfs:seeAlso ?OMA_LINK .\\n\\t\\t?protein2 orth:organism/obo:RO_0002162 ?taxon2 .\\n\\t\\t?protein2 lscr:xrefUniprot ?uniprot .\\n\\t\\tFILTER ( ?node1 != ?node2 )\\n\\t}\\n\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t?uniprot up:annotation ?annotation .\\n\\t\\t?annotation rdfs:comment ?annotation_text .\\n\\t\\t?annotation a up:Disease_Annotation .\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Which are the genes in Human associated to a disease that are orthologous to a gene expressed in the rat\\u2019s brain?\",\n",
      "    \"results\": 6\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX genex: <http://purl.org/genex#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT DISTINCT ?anat_label ?GO_CLASS ?GO_LABEL {\\n\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\tSELECT ?geneEns ?anat_label {\\n\\t\\t\\t?geneB a orth:Gene .\\n\\t\\t\\t?geneB genex:isExpressedIn ?anat .\\n\\t\\t\\t?geneB rdfs:label ?anatLabel ;\\n\\t\\t\\t\\tlscr:xrefEnsemblGene ?geneEns .\\n\\t\\t\\tFILTER (UCASE($anatLabel) = UCASE('ins'))\\n\\t\\t\\t?anat rdfs:label ?anat_label .\\n\\t\\t\\t?geneB orth:organism ?o .\\n\\t\\t\\t?o obo:RO_0002162 ?taxon2 .\\n\\t\\t\\t?taxon2 up:commonName 'zebrafish' .\\n\\t\\t}\\n\\t}\\n\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t?uniprot rdfs:seeAlso / up:transcribedFrom ?geneEns .\\n\\t\\t?uniprot a up:Protein .\\n\\t\\t?uniprot up:classifiedWith ?GO_CLASS .\\n\\t\\t?GO_CLASS rdfs:label ?GO_LABEL .\\n\\t}\\n}\\nLIMIT 10\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Anatomic entities where the ins zebrafish gene is expressed and its gene GO annotations.\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT DISTINCT ?proteinOMA ?species ?mnemonic ?evidenceType ?UniProt_URI WHERE {\\n\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\t?proteinOMA a orth:Protein ;\\n\\t\\t\\torth:organism/obo:RO_0002162/up:scientificName ?species ;\\n\\t\\t\\trdfs:label 'INS' .\\n\\t\\t?proteinOMA lscr:xrefUniprot ?UniProt_URI .\\n\\t}\\n\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t?UniProt_URI up:mnemonic ?mnemonic ;\\n\\t\\t\\tup:existence/rdfs:label ?evidenceType .\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Proteins in OMA that are encoded by the INS gene without considering synonyms and their species identifier and evidence types from UniProt database.\",\n",
      "    \"results\": 76\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT DISTINCT ?PROTEIN_1 ?PROTEIN_2 ?UNIPROT_XREF_1 ?UNIPROT_XREF_2 WHERE {\\n\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t?taxon_1 up:commonName 'Mouse' .\\n\\t\\t?taxon_2 up:commonName 'Rabbit' .\\n\\t}\\n\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t?node2 orth:hasHomologousMember* ?PROTEIN_2 .\\n\\t\\t?node1 orth:hasHomologousMember* ?PROTEIN_1 .\\n\\t\\t?PROTEIN_1 a orth:Protein .\\n\\t\\t?PROTEIN_1 orth:organism/obo:RO_0002162 ?taxon_1 ;\\n\\t\\t\\trdfs:label 'HBB-Y' ;\\n\\t\\t\\tlscr:xrefUniprot ?UNIPROT_XREF_1 .\\n\\t\\t?PROTEIN_2 a orth:Protein .\\n\\t\\t?PROTEIN_2 orth:organism/obo:RO_0002162 ?taxon_2 .\\n\\t\\t?PROTEIN_2 lscr:xrefUniprot ?UNIPROT_XREF_2 .\\n\\t\\tFILTER ( ?node1 != ?node2 )\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Rabbit's proteins encoded by genes that are orthologous to Mouse's HBB-Y gene and their cross reference links to Uniprot\",\n",
      "    \"results\": 24\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT ?name1 ?protein1 ?name2 ?protein2 ?OMA_link2 ?anatomicalEntity {\\n\\tSELECT DISTINCT * {\\n\\t\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\t\\t?taxon up:commonName 'human' ;\\n\\t\\t\\t\\tup:commonName ?name1 .\\n\\t\\t\\t?taxon2 up:commonName 'mouse' ;\\n\\t\\t\\t\\tup:commonName ?name2 .\\n\\t\\t}\\n\\t\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t\\t?node2 orth:hasHomologousMember* ?protein2 .\\n\\t\\t\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t\\t\\t?protein1 a orth:Protein .\\n\\t\\t\\t?protein1 rdfs:label 'INS' ;\\n\\t\\t\\t\\torth:organism/obo:RO_0002162 ?taxon .\\n\\t\\t\\t?protein2 a orth:Protein ;\\n\\t\\t\\t\\tsio:SIO_010079 ?gene ; #is encoded by\\n\\t\\t\\t\\torth:organism/obo:RO_0002162 ?taxon2 .\\n\\t\\t\\t?gene lscr:xrefEnsemblGene ?geneEns .\\n\\t\\t\\t?protein2 rdfs:seeAlso ?OMA_link2 .\\n\\t\\t\\tFILTER ( ?node1 != ?node2 )\\n\\t\\t}\\n\\t\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\t\\t?geneB a orth:Gene .\\n\\t\\t\\t\\t?geneB genex:isExpressedIn ?cond .\\n\\t\\t\\t\\t?cond genex:hasAnatomicalEntity ?anat .\\n\\t\\t\\t\\t?geneB lscr:xrefEnsemblGene ?geneEns .\\n\\t\\t\\t?anat rdfs:label 'liver' ;\\n\\t\\t\\t\\trdfs:label ?anatomicalEntity .\\n\\t\\t\\t?geneB orth:organism ?o .\\n\\t\\t\\t?o obo:RO_0002162 ?taxon2 .\\n\\t\\t}\\n\\t}\\n\\tLIMIT 10\\n}\\nLIMIT 10\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Which are the mouse's proteins encoded by genes which are expressed in the liver and are orthologous to human's INS gene?\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\nSELECT DISTINCT ?id ?OMA_LINK WHERE {\\n\\tSELECT * {\\n\\t\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\t\\tSELECT DISTINCT ?gene ?id {\\n\\t\\t\\t\\t?gene a orth:Gene .\\n\\t\\t\\t\\t?gene genex:isExpressedIn ?anat .\\n\\t\\t\\t\\t?anat rdfs:label 'brain' .\\n\\t\\t\\t\\t?gene orth:organism ?o .\\n\\t\\t\\t\\t?o obo:RO_0002162 ?taxon .\\n\\t\\t\\t\\t?gene dcterms:identifier ?id .\\n\\t\\t\\t\\t?taxon up:commonName 'fruit fly' .\\n\\t\\t\\t}\\n\\t\\t\\tLIMIT 100\\n\\t\\t}\\n\\t\\tSERVICE <https://sparql.omabrowser.org/lode/sparql> {\\n\\t\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t\\t?node2 orth:hasHomologousMember* ?protein2 .\\n\\t\\t\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t\\t\\t?protein1 dcterms:identifier ?id .\\n\\t\\t\\t?protein2 rdfs:seeAlso ?OMA_LINK .\\n\\t\\t\\tFILTER ( ?node1 != ?node2 )\\n\\t\\t}\\n\\t}\\n}\\nLIMIT 10\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Which are the orthologs of a gene that is expressed in the fruit fly\\u2019s brain?\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\nSELECT DISTINCT ?orthologous_to_gene_id ?orthologous_to_gene ?hasOrtholog WHERE {\\n\\tSELECT * {\\n\\t\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\t\\tSELECT DISTINCT ?gene ?orthologous_to_gene_id ?orthologous_to_gene {\\n\\t\\t\\t\\t?gene a orth:Gene .\\n\\t\\t\\t\\t?gene genex:isExpressedIn ?anat .\\n\\t\\t\\t\\t?anat rdfs:label 'eye' .\\n\\t\\t\\t\\t?gene orth:organism ?o .\\n\\t\\t\\t\\t?o obo:RO_0002162 ?taxon .\\n\\t\\t\\t\\t?gene dcterms:identifier ?orthologous_to_gene_id .\\n\\t\\t\\t\\t?gene rdfs:seeAlso ?orthologous_to_gene .\\n\\t\\t\\t\\t?taxon up:commonName 'fruit fly' .\\n\\t\\t\\t} LIMIT 100\\n\\t\\t}\\n\\t\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\t\\t?protein2 a orth:Protein .\\n\\t\\t\\t?protein1 a orth:Protein .\\n\\t\\t\\t?clusterPrimates a orth:OrthologsCluster .\\n\\t\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t\\t?node2 orth:hasHomologousMember* ?clusterPrimates .\\n\\t\\t\\t?clusterPrimates orth:hasHomologousMember* ?protein2 .\\n\\t\\t\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t\\t\\t?protein1 dcterms:identifier ?orthologous_to_gene_id .\\n\\t\\t\\t?protein2 rdfs:seeAlso ?hasOrtholog .\\n\\t\\t\\t?clusterPrimates orth:hasTaxonomicRange ?taxRange .\\n\\t\\t\\t?taxRange orth:taxRange ?taxLevel .\\n\\t\\t\\tVALUES ?taxLevel { 'Primates' }\\n\\t\\t\\tFILTER ( ?node1 != ?node2 )\\n\\t\\t}\\n\\t}\\n}\\nLIMIT 10\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Which are the genes in Primates orthologous to a gene that is expressed in the fruit fly\\u2019s eye?\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT * {\\n\\t?species a up:Taxon .\\n\\t?species up:scientificName ?name .\\n\\t?species up:rank up:Species .\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql\",\n",
      "    \"question\": \"Scientific name of species in bgee with their uniprot taxon\",\n",
      "    \"results\": 52\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\nSELECT ?geneName ?geneId {\\n\\t?gene a orth:Gene .\\n\\t?gene rdfs:label ?geneName .\\n\\t?gene dcterms:identifier ?geneId .\\n\\t?gene dcterms:description ?desc .\\n\\tFILTER CONTAINS ( ?desc, 'lung' )\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql\",\n",
      "    \"question\": \"Genes with the following term in the description: lung\",\n",
      "    \"results\": 34\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT ?name {\\n\\t?gene a orth:Gene .\\n\\t?gene rdfs:label ?geneName .\\n\\t?gene orth:organism ?organism . #orth v2\\n\\t?organism obo:RO_0002162 ?taxon . #label: in taxon .\\n\\t?taxon up:scientificName ?name .\\n\\tFILTER ( UCASE(?geneName) = UCASE('Mt-co1') )\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql\",\n",
      "    \"question\": \"Which species the Mt-co1 gene is present (without considering synonyms)?\",\n",
      "    \"results\": 11\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT distinct ?gene_page ?geneName {\\n\\t?gene a orth:Gene .\\n\\t?gene rdfs:seeAlso ?gene_page .\\n\\t?gene rdfs:label ?geneName .\\n\\t?gene orth:organism ?taxon .\\n\\t?taxon obo:RO_0002162 ?species . #in taxon property .\\n\\t?species up:commonName ?commonName .\\n\\tFILTER ( lcase(str(?commonName)) = lcase(\\\"Fruit fly\\\") ) .\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql\",\n",
      "    \"question\": \"List the labels and identifiers of Fruit fly genes in bgee\",\n",
      "    \"results\": 17807\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT DISTINCT ?anatName {\\n\\t?cond genex:hasAnatomicalEntity ?anatEntity .\\n\\t?anatEntity rdfs:label ?anatName .\\n\\t?cond genex:hasDevelopmentalStage ?stage .\\n\\t?stage rdfs:label ?stageName .\\n\\t?cond obo:RO_0002162 ?taxon .\\n\\t?taxon up:commonName ?commonName .\\n\\tFILTER ( lcase(str(?commonName)) = lcase(\\\"Human\\\") ).\\n\\tFILTER ( CONTAINS(?stageName, \\\"young adult\\\") )\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql\",\n",
      "    \"question\": \"Human anatomic entities at young adult developmental stage\",\n",
      "    \"results\": 427\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT ?anatEntity ?anatName {\\n\\t?seq a orth:Gene .\\n\\t?seq genex:isExpressedIn ?cond.\\n\\t?seq rdfs:label 'apoc1' .\\n\\t?cond genex:hasAnatomicalEntity ?anatEntity .\\n\\t?anatEntity rdfs:label ?anatName . \\n} LIMIT 10\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql\",\n",
      "    \"question\": \"Anatomic entities where the apoc1 gene is expressed\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT DISTINCT ?anatEntity ?anatName {\\n\\t?seq a orth:Gene .\\n\\t?seq genex:isExpressedIn ?cond .\\n\\t?seq rdfs:label ?geneName .\\n\\t?expr genex:hasExpressionCondition ?cond .\\n\\t?cond genex:hasAnatomicalEntity ?anatEntity .\\n\\t?anatEntity rdfs:label ?anatName .\\n\\t?cond obo:RO_0002162 <http://purl.uniprot.org/taxonomy/10116> . \\n\\tFILTER (LCASE(?geneName) = LCASE('apoc1') )\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql\",\n",
      "    \"question\": \"Anatomic entities in rat where the apoc1 gene is expressed\",\n",
      "    \"results\": 64\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT DISTINCT ?anatEntity ?anatName {\\n\\t?seq a orth:Gene .\\n\\t?seq genex:isExpressedIn ?cond .\\n\\t?seq rdfs:label ?geneName .\\n\\t?cond genex:hasAnatomicalEntity ?anatEntity .\\n\\t?anatEntity rdfs:label ?anatName .\\n\\t?cond genex:hasDevelopmentalStage ?stage .\\n\\t?stage rdfs:label\\n\\t?stageName .\\n\\t?cond obo:RO_0002162 ?taxon . #in taxon property .\\n\\t?taxon up:commonName ?commonName .\\n\\tFILTER ( lcase(str(?commonName)) = lcase(\\\"Mouse\\\") ) .\\n\\tFILTER ( CONTAINS(?stageName,'adult') && LCASE(?geneName) = LCASE('apoc1') )\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql\",\n",
      "    \"question\": \"Anatomic entities at the Mouse\\u2019s adult stage where the apoc1 gene is expressed\",\n",
      "    \"results\": 335\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT * WHERE {\\n\\t?s a up:Taxon .\\n\\t?s up:scientificName ?v .\\n\\t?s up:rank up:Species .\\n} \",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Which species are available on OMA database and their scientific name?\",\n",
      "    \"results\": 2851\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein ?OMA_link WHERE {\\n\\t?protein a orth:Protein .\\n\\t?protein orth:organism ?organism .\\n\\t?inTaxon rdfs:label 'in taxon'@en .\\n\\t?organism ?inTaxon ?taxon .\\n\\t?taxon up:scientificName 'Rattus norvegicus' .\\n\\t?protein rdfs:seeAlso ?OMA_link .\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Find Rattus norvegicus proteins present in OMA database\",\n",
      "    \"results\": 22232\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?taxon ?target (COUNT(?protein) as ?Protein_count) WHERE {\\n\\t?protein a orth:Protein .\\n\\t?protein orth:organism ?organism .\\n\\t?inTaxon rdfs:label 'in taxon'@en .\\n\\t?organism ?inTaxon ?taxon .\\n\\t?taxon up:scientificName ?target .\\n}\\nGROUP BY ?taxon ?target\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"How many proteins (and their scientific name) per species are present in OMA database?\",\n",
      "    \"results\": 2851\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?OMA_LINK WHERE {\\n\\t#The three that contains Orthologs. The leafs are proteins .\\n\\t#This graph pattern defines the relationship protein1 is Orthologs to protein2\\n\\t?cluster a orth:OrthologsCluster .\\n\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t?node2 orth:hasHomologousMember* ?protein2 .\\n\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t########\\n\\t#Specify the protein to look for its orthologs\\n\\t?protein1 sio:SIO_010079 ?gene .\\n\\t?gene lscr:xrefEnsemblGene ensembl:ENSLACG00000002497 .\\n\\t########\\n\\t#The OMA link to the second protein\\n\\t?protein2 rdfs:seeAlso ?OMA_LINK .\\n\\t########\\n\\tFILTER(?node1 != ?node2)\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"All proteins that are orthologous to ENSLACG00000002497 Ensembl gene (identifier) \",\n",
      "    \"results\": 42\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?OMA_LINK WHERE {\\n\\t#The three that contains paralogs. The leafs are proteins .\\n\\t#This graph pattern defines the relationship protein1 is paralogous to protein2\\n\\t?cluster a orth:ParalogsCluster .\\n\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t?node2 orth:hasHomologousMember* ?protein2 .\\n\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t########\\n\\t#Specify the protein to look for its paralogs\\n\\t?protein1 sio:SIO_010079 ?gene .\\n\\t?gene lscr:xrefEnsemblGene ensembl:ENSG00000244734 .\\n\\t########\\n\\t#The OMA link to the second protein\\n\\t?protein2 rdfs:seeAlso ?OMA_LINK .\\n\\t########\\n\\tFILTER(?node1 != ?node2)\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Proteins that are paralogous to ENSG00000244734 Ensembl gene (identifier)\",\n",
      "    \"results\": 473\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX dct: <http://purl.org/dc/terms/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT ?protein1 ?protein2 ?geneName2 ?species2 ?OMA_LINK2 ?Prot2_uniprot ?prot2_ensemblGeneId WHERE {\\n\\t?cluster a orth:OrthologsCluster .\\n\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t?node2 orth:hasHomologousMember* ?protein2 .\\n\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t?protein1 a orth:Protein ;\\n\\t\\torth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus' ;\\n\\t\\trdfs:label 'APOCI' .\\n\\t?protein2 a orth:Protein ;\\n\\t\\trdfs:seeAlso ?OMA_LINK2 ;\\n\\t\\tlscr:xrefUniprot ?Prot2_uniprot ;\\n\\t\\tsio:SIO_010079/lscr:xrefEnsemblGene/dct:identifier ?prot2_ensemblGeneId ;\\n\\t\\trdfs:label ?geneName2 ;\\n\\t\\torth:organism/obo:RO_0002162/up:scientificName ?species2 .\\n\\tFILTER(?node1 != ?node2)\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"All proteins per species that are orthologous to Oryctolagus cuniculus APOCI related protein and their cross reference links to OMA and Uniprot including the corresponding Ensembl gene identifier.\\u00a0\",\n",
      "    \"results\": 56\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX dct: <http://purl.org/dc/terms/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT DISTINCT ?PROT1 ?PROT2 ?PROT1_UNIPROT_XREF ?PROT2_UNIPROT_XREF WHERE {\\n\\t?cluster a orth:OrthologsCluster .\\n\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t?node2 orth:hasHomologousMember* ?PROT2 .\\n\\t?node1 orth:hasHomologousMember* ?PROT1 .\\n\\t?PROT1 a orth:Protein .\\n\\t?PROT1 orth:organism/obo:RO_0002162/up:scientificName 'Mus musculus' ;\\n\\t\\trdfs:label 'HBB-Y' ;\\n\\t\\tlscr:xrefUniprot ?PROT1_UNIPROT_XREF .\\n\\t?PROT2 a orth:Protein .\\n\\t?PROT2 orth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus' .\\n\\t?PROT2 lscr:xrefUniprot ?PROT2_UNIPROT_XREF .\\n\\tFILTER( ?node1 != ?node2 )\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Oryctolagus cuniculus proteins (PROT1) encoded by genes that are orthologous to Mus musculus (PROT2) HBB-Y gene and their cross reference links to Uniprot\",\n",
      "    \"results\": 24\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX oma: <http://omabrowser.org/ontology/oma#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT DISTINCT ?protein1 ?protein2 {\\n\\tVALUES (?protein1){\\n\\t\\t(<http://purl.uniprot.org/uniprot/P68871>)\\n\\t}\\n\\t?cluster a orth:OrthologsCluster.\\n\\t?cluster orth:hasHomologousMember ?node1.\\n\\t?cluster orth:hasHomologousMember ?node2.\\n\\t?node1 orth:hasHomologousMember* ?protein_OMA_1.\\n\\t?node2 orth:hasHomologousMember* ?protein_OMA_2.\\n\\t?protein_OMA_1 lscr:xrefUniprot ?protein1.\\n\\t?protein_OMA_2 lscr:xrefUniprot ?protein2.\\n\\tFILTER(?node1 != ?node2)\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Retrieve pairwise orthologs which are derived from the HOGs in the OMA database of gene with the P68871 UniProt accession number\",\n",
      "    \"results\": 186\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX mbgdr: <http://mbgd.genome.ad.jp/rdf/resource/>\\nPREFIX mbgd: <http://purl.jp/bio/11/mbgd#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nSELECT ?protein1 ?protein2 WHERE {\\n\\tVALUES (?protein1){\\n\\t\\t(<http://purl.uniprot.org/uniprot/P68871>)\\n\\t}\\n\\t?cluster a orth:OrthologsCluster.\\n\\t?cluster orth:hasHomologous ?node1.\\n\\t?cluster orth:hasHomologous ?node2.\\n\\t?node1 orth:hasHomologous* ?gene1.\\n\\t?node2 orth:hasHomologous* ?gene2.\\n\\t?gene1 mbgd:uniprot ?protein1.\\n\\t?gene2 mbgd:uniprot ?protein2.\\n\\tFILTER(?node1 != ?node2)\\n}\",\n",
      "    \"endpoint\": \"http://sparql.nibb.ac.jp/sparql\",\n",
      "    \"question\": \"Retrieve pairwise orthologs which are derived from the HOGs in the MBGD database of gene with the P68871 UniProt accession number - only works if you disable CORS in your browser\",\n",
      "    \"results\": 18\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nSELECT DISTINCT ?cluster ?protein2_OMA_URI ?protein2_uniprot_URI ?tax_name {\\n\\tVALUES (?protein1_uniprot_URI){\\n\\t\\t(<http://purl.uniprot.org/uniprot/P68871>)\\n\\t}\\n\\tVALUES (?tax_name){\\n\\t\\t(\\\"Primates\\\")\\n\\t}\\n\\t?cluster a orth:OrthologsCluster.\\n\\t?cluster orth:hasHomologousMember* ?protein_OMA_1.\\n\\t?cluster orth:hasHomologousMember* ?protein2_OMA_URI.\\n\\t?protein_OMA_1 a orth:Protein.\\n\\t?protein2_OMA_URI a orth:Protein.\\n\\t?protein_OMA_1 lscr:xrefUniprot ?protein1_uniprot_URI.\\n\\tOPTIONAL {\\n\\t\\t?protein2_OMA_URI lscr:xrefUniprot ?protein2_uniprot_URI.\\n\\t}\\n\\t?cluster orth:hasTaxonomicRange ?tax.\\n\\t?tax orth:taxRange ?tax_name.\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Retrieve homologous groups (i.e. clusters) which are derived from the HOGs in the OMA database that contain the gene with the P68871 UniProt accession number\",\n",
      "    \"results\": 55\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT DISTINCT ?root_hog ?species_name ?protein1_uniprot (?protein1 as ?protein1_OMA) ?taxLevel {\\n\\tVALUES ?protein2_uniprot {\\n\\t\\t<http://purl.uniprot.org/uniprot/P68871>\\n\\t}\\n\\t?root_hog obo:CDAO_0000148 ?hog_cluster. #has_Root\\n\\t?hog_cluster orth:hasHomologousMember* ?node1.\\n\\t?node1 a orth:OrthologsCluster.\\n\\t?node1 orth:hasTaxonomicRange ?level.\\n\\t?level orth:taxRange ?taxLevel.\\n\\t?node1 orth:hasHomologousMember* ?protein1.\\n\\t?hog_cluster orth:hasHomologousMember* ?protein2.\\n\\t?protein1 a orth:Protein.\\n\\t?protein1 orth:organism ?organism.\\n\\t?organism obo:RO_0002162 ?taxon.\\n\\t?taxon up:scientificName ?species_name.\\n\\tOPTIONAL {\\n\\t\\t?protein1 lscr:xrefUniprot ?protein1_uniprot\\n\\t}\\n\\t?protein2 a orth:Protein.\\n\\t?protein2 lscr:xrefUniprot ?protein2_uniprot.\\n} ORDER BY ?taxLevel\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Retrieve hierarchical orthologous groups from the OMA database that contain the gene with the P68871 UniProt accession number\",\n",
      "    \"results\": 11125\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX orthodb: <http://purl.orthodb.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT DISTINCT ?hog_root ?species_name ?protein1_uniprot ?gene1 ?taxLevel_uniprot ?taxLevel WHERE {\\n\\tVALUES ?protein2_uniprot {\\n\\t\\t<http://purl.uniprot.org/uniprot/P68871>\\n\\t}\\n\\t?gene2 a orthodb:Gene.\\n\\t?gene2 orthodb:memberOf ?groups.\\n\\t?gene2 orthodb:memberOf ?hog_root.\\n\\tFILTER NOT EXISTS {\\n\\t\\t?hog_root orthodb:ancestralOG ?ancestor.\\n\\t}\\n\\t?groups orthodb:ancestralOG* ?hog_root.\\n\\t?gene1 a orthodb:Gene.\\n\\t?gene1 orthodb:memberOf ?groups.\\n\\t?gene1 up:organism ?organism.\\n\\t?organism a ?taxon.\\n\\t?taxon up:scientificName ?species_name.\\n\\t?groups orthodb:ogBuiltAt ?taxLevel_uniprot.\\n\\t?taxLevel_uniprot up:scientificName ?taxLevel.\\n\\t?gene2 orthodb:xref ?xref2.\\n\\t?xref2 orthodb:xrefResource ?protein2_uniprot.\\n\\t?protein2_uniprot a orthodb:Uniprot.\\n\\t?gene1 orthodb:xref ?xref.\\n\\t?xref orthodb:xrefResource ?protein1_uniprot.\\n\\t?protein1_uniprot a orthodb:Uniprot.\\n} ORDER BY ?hog_root ?taxLevel\",\n",
      "    \"endpoint\": \"https://sparql.orthodb.org/sparql\",\n",
      "    \"question\": \"Retrieve hierarchical orthologous groups from the OrthoDB that contain the gene with the P68871 UniProt accession number - only works if you disable CORS in your browser\",\n",
      "    \"results\": 3824\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX oma: <http://omabrowser.org/ontology/oma#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX mbgd: <http://purl.jp/bio/11/mbgd#>\\nSELECT ?protein2 ?species WHERE {\\n\\tSERVICE<http://sparql.nibb.ac.jp/sparql> {\\n\\t\\tSELECT ?protein2 ?species where {\\n\\t\\t\\t?cluster_mbgd a orth:OrthologsCluster.\\n\\t\\t\\t?cluster_mbgd orth:hasHomologous ?node1_mbgd.\\n\\t\\t\\t?cluster_mbgd orth:hasHomologous ?node2_mbgd.\\n\\t\\t\\t?node1_mbgd orth:hasHomologous* ?gene1.\\n\\t\\t\\t?node2_mbgd orth:hasHomologous* ?gene2.\\n\\t\\t\\t?gene1 mbgd:uniprot <http://purl.uniprot.org/uniprot/K9Z723>.\\n\\t\\t\\t?gene2 mbgd:uniprot ?protein2.\\n\\t\\t\\t?gene2 mbgd:organism ?taxon.\\n\\t\\t\\tOPTIONAL {\\n\\t\\t\\t\\t?taxon mbgd:species ?species.\\n\\t\\t\\t}\\n\\t\\t\\tFILTER (?node1_mbgd != ?node2_mbgd)\\n\\t\\t}\\n\\t}\\n\\tFILTER NOT EXISTS { # keep only those that do not exist in OMA\\n\\t\\t?cluster a orth:OrthologsCluster.\\n\\t\\t?cluster orth:hasHomologousMember ?node1.\\n\\t\\t?cluster orth:hasHomologousMember ?node2.\\n\\t\\t?node1 orth:hasHomologousMember* ?protein_OMA_1.\\n\\t\\t?node2 orth:hasHomologousMember* ?protein_OMA_2.\\n\\t\\t?protein_OMA_1 lscr:xrefUniprot <http://purl.uniprot.org/uniprot/K9Z723>.\\n\\t\\t?protein_OMA_2 lscr:xrefUniprot ?protein2.\\n\\t\\tFILTER (?node1 != ?node2)\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Retrieve pairwise orthologous genes of the Cyanobacterium-aponinum psb27- gene that are found in the MBGD database but are not present in OMA\",\n",
      "    \"results\": 104\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein WHERE\\n{\\n\\t?protein a up:Protein .\\n\\t?protein up:mnemonic 'A4_HUMAN'\\n}\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select the UniProt protein with the mnemonic 'A4_HUMAN'\",\n",
      "    \"results\": 1\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein WHERE {\\n\\t?protein a up:Protein .\\n\\t?protein up:created '2010-11-30'^^xsd:date\\n}\\nLIMIT 10\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select all UniProt proteins that were integrated on the 2010-11-30\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT * WHERE {\\n\\t?target a up:Taxon .\\n}\\nLIMIT 10\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select all taxa from the UniProt taxonomy\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?taxon ?target WHERE {\\n\\t?taxon a up:Taxon .\\n\\t?taxon up:scientificName ?target .\\n\\t?taxon rdfs:subClassOf taxon:2 .\\n}\\nLIMIT 10\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select all bacterial taxa, and their up:scientificName from the UniProt taxonomy\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT * WHERE {\\n\\t?protein a up:Protein .\\n\\t?protein up:mnemonic 'A4_HUMAN' .\\n\\t?protein rdfs:seeAlso ?ensembl_Transcript .\\n\\t?ensembl_Transcript a up:Transcript_Resource .\\n\\t?ensembl_Transcript up:transcribedFrom ?ensembl_Gene .\\n\\t?ensembl_Transcript up:database <http://purl.uniprot.org/database/Ensembl> .\\n}\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"which are the Ensembl transcript and genes for the A4_HUMAN protein \",\n",
      "    \"results\": 6\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?protein ?annotation WHERE {\\n\\t?protein a up:Protein .\\n\\t?protein up:organism taxon:9606 .\\n\\t?protein up:annotation ?annotation .\\n\\t?annotation a up:Natural_Variant_Annotation .\\n\\t?annotation rdfs:comment ?text .\\n\\tFILTER ( CONTAINS(?text,'loss of function') )\\n}\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select all human UniProt entries with a sequence variant that leads to a loss of function\",\n",
      "    \"results\": 782\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nSELECT ?name ?diseaseAnnotation WHERE {\\n\\t?protein a up:Protein .\\n\\t?protein up:organism taxon:9606 .\\n\\t?protein up:encodedBy ?gene .\\n\\t?gene skos:prefLabel ?name .\\n\\t?protein up:annotation ?annotation .\\n\\t?annotation a up:Disease_Annotation .\\n\\t?annotation rdfs:comment ?diseaseAnnotation .\\n}\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select the preferred gene name and disease annotation of all human UniProt proteins that are known to be involved in a disease\",\n",
      "    \"results\": 31652\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT ?protein ?begin ?end WHERE {\\n\\t?protein a up:Protein .\\n\\t?protein up:annotation ?annotation .\\n\\t?annotation a up:Transmembrane_Annotation .\\n\\t?annotation up:range ?range .\\n\\t?range faldo:begin/faldo:position ?begin .\\n\\t?range faldo:end/faldo:position ?end .\\n} \\nLIMIT 10\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select all UniProt proteins with annotated transmembrane regions and the regions' begin and end coordinates on the canonical sequence\",\n",
      "    \"results\": 10\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX ec: <http://purl.uniprot.org/enzyme/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?enzyme (COUNT(?protein) as ?number) WHERE {\\n\\t?protein up:enzyme ?enzyme .\\n\\t?enzyme rdfs:subClassOf ec:1.1.1.- .\\n}\\nGROUP BY ?enzyme ORDER BY ?enzyme\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select the number of proteins for each of the subclasses of ec:1.1.1.- EC (Enzyme Commission) class\",\n",
      "    \"results\": 358\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?protein ?aa WHERE {\\n\\t?protein a up:Protein .\\n\\t?protein up:organism ?organism .\\n\\t{\\n\\t\\t?protein up:organism taxon:8333 .\\n\\t}\\n\\tUNION\\n\\t{\\n\\t\\t?protein up:organism ?organism .\\n\\t\\t?organism rdfs:subClassOf taxon:8333 .\\n\\t}\\n\\t?protein up:sequence ?s .\\n\\t?s rdf:value ?aa\\n}\",\n",
      "    \"endpoint\": \"https://sparql.uniprot.org/sparql\",\n",
      "    \"question\": \"Select all\\u00a0E-Coli K12 (including strains)\\u00a0UniProt proteins and their amino acid sequence\",\n",
      "    \"results\": 3152\n",
      "  }\n",
      "]\n",
      "[\n",
      "  {\n",
      "    \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nSELECT distinct ?gene_id ?orthologousTo ?uniprot ?annotation_text WHERE {\\n\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\tSELECT ?gene ?gene_id {\\n\\t\\t\\t?gene a orth:Gene .\\n\\t\\t\\t?gene genex:isExpressedIn ?anat .\\n\\t\\t\\t?gene dcterms:identifier ?gene_id .\\n\\t\\t\\t?anat rdfs:label 'brain' .\\n\\t\\t\\t?gene orth:organism ?o .\\n\\t\\t\\t?o obo:RO_0002162 ?taxon .\\n\\t\\t\\t?taxon up:commonName \\\"fruit fly\\\" .\\n\\t\\t}\\n\\t\\tLIMIT 10\\n\\t}\\n\\n\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\tSELECT ?uniprot ?gene ?orthologousTo {\\n\\t\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t\\t?node2 orth:hasHomologousMember* ?orthologousTo .\\n\\t\\t\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t\\t\\t?protein1 sio:SIO_010079 ?gene .\\n\\t\\t\\t?orthologousTo lscr:xrefUniprot ?uniprot .\\n\\t\\t\\tfilter(?node1 != ?node2)\\n\\t\\t}\\n\\t}\\n\\n\\tSERVICE <http://sparql.uniprot.org/sparql>{\\n\\t\\t?uniprot up:annotation ?annotation .\\n\\t\\t?annotation rdfs:comment ?annotation_text .\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"The orthologs of a gene that is expressed in the fruit fly\\u2019s brain and the UniProt annotations of these orthologs.\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nSELECT ?gene ?OMA_LINK ?uniprot ?annotation_text WHERE {\\n\\tSELECT * {\\n\\t\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\t\\tSELECT ?gene {\\n\\t\\t\\t\\t?gene genex:isExpressedIn ?anat .\\n\\t\\t\\t\\t?anat rdfs:label 'brain' .\\n\\t\\t\\t\\t?gene orth:organism ?o .\\n\\t\\t\\t\\t?o obo:RO_0002162 ?taxon .\\n\\t\\t\\t\\t?taxon up:commonName 'fruit fly' .\\n\\t\\t\\t}\\n\\t\\t\\tLIMIT 10\\n\\t\\t}\\n\\t\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\t\\tSELECT ?gene ?OMA_LINK ?uniprot{\\n\\t\\t\\t\\t?protein2 a orth:Protein .\\n\\t\\t\\t\\t?protein1 a orth:Protein .\\n\\t\\t\\t\\t?clusterPrimates a orth:OrthologsCluster .\\n\\t\\t\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t\\t\\t?node2 orth:hasHomologousMember* ?clusterPrimates .\\n\\t\\t\\t\\t?clusterPrimates orth:hasHomologousMember* ?protein2 .\\n\\t\\t\\t\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t\\t\\t\\t?protein1 sio:SIO_010079 ?gene. #encoded by\\n\\t\\t\\t\\t?protein2 rdfs:seeAlso ?OMA_LINK .\\n\\t\\t\\t\\t?clusterPrimates orth:hasTaxonomicRange ?taxRange .\\n\\t\\t\\t\\t?taxRange orth:taxRange ?taxLevel .\\n\\t\\t\\t\\t?protein2 lscr:xrefUniprot ?uniprot .\\n\\t\\t\\t\\tVALUES ?taxLevel { 'Primates' }\\n\\t\\t\\t\\tFILTER ( ?node1 != ?node2 )\\n\\t\\t\\t}\\n\\t\\t}\\n\\t\\tSERVICE <https://sparql.uniprot.org/sparql> {\\n\\t\\t\\tSELECT ?annotation_text ?uniprot {\\n\\t\\t\\t\\t?uniprot up:annotation ?annotation .\\n\\t\\t\\t\\t?annotation rdfs:comment ?annotation_text .\\n\\t\\t\\t}\\n\\t\\t}\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Which are the genes in Primates orthologous to a gene that is expressed in the fruit fly\\u2019s brain and the UniProt annotations of the genes.\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\nSELECT DISTINCT ?protein ?orthologous_protein ?gene ?annotation_text WHERE {\\n\\tSELECT * {\\n\\t\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t\\tSELECT ?protein ?annotation_text WHERE {\\n\\t\\t\\t\\t?protein a up:Protein ;\\n\\t\\t\\t\\tup:organism taxon:9606 ;\\n\\t\\t\\t\\tup:annotation ?annotation .\\n\\t\\t\\t\\t?annotation rdfs:comment ?annotation_text .\\n\\t\\t\\t\\t?annotation a up:Disease_Annotation .\\n\\t\\t\\t\\tFILTER CONTAINS (?annotation_text, \\\"glioblastoma\\\")\\n\\t\\t\\t}\\n\\t\\t}\\n\\t\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\t\\tSELECT ?orthologous_protein ?protein ?gene WHERE {\\n\\t\\t\\t\\t?protein_OMA a orth:Protein .\\n\\t\\t\\t\\t?orthologous_protein a orth:Protein .\\n\\t\\t\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t\\t\\t?cluster\\n\\t\\t\\t\\torth:hasHomologousMember ?node2 .\\n\\t\\t\\t\\t?node2 orth:hasHomologousMember* ?protein_OMA .\\n\\t\\t\\t\\t?node1 orth:hasHomologousMember* ?orthologous_protein .\\n\\t\\t\\t\\t?orthologous_protein orth:organism/obo:RO_0002162 taxon:10116 . # rattus norvegicus\\n\\t\\t\\t\\t?orthologous_protein sio:SIO_010079 ?gene .\\n\\t\\t\\t\\t?protein_OMA lscr:xrefUniprot ?protein .\\n\\t\\t\\t\\tFILTER(?node1 != ?node2)\\n\\t\\t\\t}\\n\\t\\t}\\n\\t\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\t\\t ?gene genex:isExpressedIn ?a .\\n\\t\\t\\t?a rdfs:label \\\"brain\\\" .\\n\\t\\t\\t?gene orth:organism ?s . \\n\\t\\t\\t?s obo:RO_0002162 taxon:10116.\\n\\t\\t}\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Which are the proteins associated with \\\"glioblastoma\\\" and the orthologs expressed in the rat\\u2019s brain?\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up:<http://purl.uniprot.org/core/>\\nPREFIX taxon:<http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth:<http://purl.org/net/orth#>\\nPREFIX dcterms:<http://purl.org/dc/terms/>\\nPREFIX obo:<http://purl.obolibrary.org/obo/>\\nPREFIX lscr:<http://purl.org/lscr#>\\nPREFIX genex:<http://purl.org/genex#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\n\\nSELECT ?gene ?orthologous_protein2 WHERE {\\n\\tSELECT * {\\n\\t\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t\\tSELECT ?protein1 WHERE {\\n\\t\\t\\t?protein1 a up:Protein;\\n\\t\\t\\t\\tup:organism/up:scientificName 'Homo sapiens' ;\\n\\t\\t\\t\\tup:annotation ?annotation .\\n\\t\\t\\t\\t?annotation rdfs:comment ?annotation_text.\\n\\t\\t\\t\\t?annotation a up:Disease_Annotation .\\n\\t\\t\\t\\tFILTER CONTAINS (?annotation_text, \\\"cancer\\\")\\n\\t\\t\\t}\\n\\t\\t}\\n\\n\\t\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\t\\tSELECT ?orthologous_protein2 ?protein1 ?gene WHERE {\\n\\t\\t\\t\\t?protein_OMA a orth:Protein .\\n\\t\\t\\t\\t?orthologous_protein2 a orth:Protein .\\n\\t\\t\\t\\t?cluster a orth:OrthologsCluster .\\n\\t\\t\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t\\t\\t?node2 orth:hasHomologousMember* ?protein_OMA .\\n\\t\\t\\t\\t?node1 orth:hasHomologousMember* ?orthologous_protein2 \\n\\t\\t\\t\\t.?orthologous_protein2 orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus' .\\n\\t\\t\\t\\t?orthologous_protein2 sio:SIO_010079 ?gene .\\n\\t\\t\\t\\t?protein_OMA lscr:xrefUniprot ?protein1 .\\n\\t\\t\\t\\tFILTER(?node1 != ?node2)\\n\\t\\t\\t}\\n\\t\\t}\\n\\n\\t\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\t\\t?gene genex:isExpressedIn ?anatEntity .\\n\\t\\t\\t?anatEntity rdfs:label 'brain' .\\n\\t\\t\\t?gene orth:organism ?org . \\n\\t\\t\\t?org obo:RO_0002162 taxon:10116 .\\n\\t\\t}\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"What are the Homo sapiens\\u2019 genes associated with cancer and their orthologs expressed in the Rattus norvegicus\\u2019s brain?\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nSELECT DISTINCT ?geneEns ?uniprot ?annotation_text {\\n\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\tSELECT ?geneEns {\\n\\t\\t\\t?geneB a orth:Gene .\\n\\t\\t\\t?geneB genex:isExpressedIn ?cond .\\n\\t\\t\\t?cond genex:hasAnatomicalEntity ?anat .\\n\\t\\t\\t\\t?geneB lscr:xrefEnsemblGene ?geneEns .\\n\\t\\t?anat rdfs:label 'pancreas' .\\n\\t\\t?geneB orth:organism ?o .\\n\\t\\t\\t?o obo:RO_0002162 ?taxon2 .\\n\\t\\t\\t?taxon2 up:commonName 'human' .\\n\\t\\t}\\n\\t\\tLIMIT 10\\n\\t}\\n\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t?uniprot rdfs:seeAlso / up:transcribedFrom ?geneEns .\\n\\t\\t?uniprot up:annotation ?annotation .\\n\\t\\t?annotation rdfs:comment ?annotation_text .\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Genes expressed in the human's pancreas and their annotations in UniProt.\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nSELECT DISTINCT ?geneEns ?uniprot ?annotation_text {\\n\\tSERVICE <https://www.bgee.org/sparql/> {\\n\\t\\tSELECT ?geneEns {\\n\\t\\t\\t?geneB genex:isExpressedIn ?cond ;\\n\\t\\t\\t\\tlscr:xrefEnsemblGene ?geneEns .\\n\\t\\t\\t?cond genex:hasDevelopmentalStage ?st .\\n\\t\\t\\t?cond genex:hasAnatomicalEntity ?anat .\\n\\t\\t\\t?st rdfs:label 'infant stage' .\\n\\t\\t\\t?anat rdfs:label 'brain' .\\n\\t\\t\\t?geneB orth:organism ?o .\\n\\t\\t\\t?o obo:RO_0002162 ?taxon2 .\\n\\t\\t\\t?taxon2 up:commonName 'human' .\\n\\t\\t}\\n\\t\\tLIMIT 10\\n\\t}\\n\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t?uniprot rdfs:seeAlso / up:transcribedFrom ?geneEns .\\n\\t\\t?uniprot up:annotation ?annotation .\\n\\t\\t?annotation a up:Disease_Annotation .\\n\\t\\t?annotation rdfs:comment ?annotation_text .\\n\\t}\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Genes expressed in the human's brain during the infant stage and their UniProt disease annotations\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT DISTINCT ?PROTEIN ?IS_PARALOGOUS_TO_PROTEIN ?UNIPROT_XREF ?PARALOG_UNIPROT_XREF ?annotation_text WHERE {\\n\\tSERVICE <https://sparql.omabrowser.org/sparql/> {\\n\\t\\t?cluster a orth:ParalogsCluster .\\n\\t\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t\\t?node2 orth:hasHomologousMember* ?PROTEIN .\\n\\t\\t?node1 orth:hasHomologousMember* ?IS_PARALOGOUS_TO_PROTEIN .\\n\\t\\t?PROTEIN a orth:Protein .\\n\\t\\t?PROTEIN orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus' ;\\n\\t\\t\\trdfs:label 'TP53' ;\\n\\t\\t\\tlscr:xrefUniprot ?UNIPROT_XREF .\\n\\t\\t?IS_PARALOGOUS_TO_PROTEIN a orth:Protein .\\n\\t\\t?IS_PARALOGOUS_TO_PROTEIN orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus' .\\n\\t\\t?IS_PARALOGOUS_TO_PROTEIN lscr:xrefUniprot ?PARALOG_UNIPROT_XREF .\\n\\t}\\n\\tSERVICE <http://sparql.uniprot.org/sparql> {\\n\\t\\t?PARALOG_UNIPROT_XREF up:annotation ?annotation .\\n\\t\\t?annotation a up:Function_Annotation .\\n\\t\\t?annotation rdfs:comment ?annotation_text .\\n\\t}\\n\\tFILTER ( ?node1 != ?node2 )\\n}\",\n",
      "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
      "    \"question\": \"Rattus norvegicus' proteins encoded by genes that are paralogous to its TP53 gene and their Uniprot function annotations.\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX dct: <http://purl.org/dc/terms/>\\nSELECT ?protein2 ?Oma_link ?Uniprot_link WHERE {\\n\\t?cluster a orth:ParalogsCluster .\\n\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t?node2 orth:hasHomologousMember* ?protein2 .\\n\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t?protein1 a orth:Protein .\\n\\t?protein1 dct:identifier 'HUMAN04027' .\\n\\t?protein2 a orth:Protein .\\n\\t?protein2 lscr:xrefUniprot ?Uniprot_link .\\n\\t?protein2 rdfs:seeAlso ?Oma_link .\\n\\tFILTER( ?node1 != ?node2 )\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Proteins that are paralogous to HUMAN04027 OMA protein (identifier) and their cross reference links to OMA and Uniprot.\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX dct: <http://purl.org/dc/terms/>\\nSELECT ?protein2 ?Oma_link ?Uniprot_link WHERE {\\n\\t?cluster a orth:OrthologsCluster .\\n\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t?node2 orth:hasHomologousMember* ?protein2 .\\n\\t?node1 orth:hasHomologousMember* ?protein1 .\\n\\t?protein1 a orth:Protein .\\n\\t?protein1 dct:identifier 'HUMAN22168' .\\n\\t?protein2 a orth:Protein .\\n\\t?protein2 lscr:xrefUniprot ?Uniprot_link .\\n\\t?protein2 rdfs:seeAlso ?Oma_link .\\n\\tFILTER(?node1 != ?node2)\\n}\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"Proteins that are orthologous to HUMAN22168 OMA protein (identifier) and their cross reference links to OMA and Uniprot.\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT (COUNT(DISTINCT ?PROTEIN) as ?num_paralogy ) ?total ((xsd:float(?num_paralogy)*100/xsd:float(?total)) as ?result) WHERE {\\n\\t?cluster a orth:ParalogsCluster .\\n\\t?cluster orth:hasHomologousMember ?node1 .\\n\\t?cluster orth:hasHomologousMember ?node2 .\\n\\t?node2 orth:hasHomologousMember* ?PROTEIN .\\n\\t?node1 orth:hasHomologousMember* ?IS_PARALOGOUS_TO_PROTEIN .\\n\\t?PROTEIN a orth:Protein .\\n\\t?PROTEIN\\north:organism/obo:RO_0002162/up:scientificName 'Drosophila melanogaster' .\\n\\t?IS_PARALOGOUS_TO_PROTEIN a orth:Protein .\\n\\t?IS_PARALOGOUS_TO_PROTEIN orth:organism/obo:RO_0002162/up:scientificName 'Drosophila melanogaster' .\\n\\t{\\n\\t\\tSELECT (COUNT(DISTINCT ?protein_total) as ?total) WHERE {\\n\\t\\t\\t?protein_total a orth:Protein .\\n\\t\\t\\t?protein_total orth:organism/obo:RO_0002162/up:scientificName 'Drosophila melanogaster' .\\n\\t\\t}\\n\\t}\\n\\tFILTER(?node1 != ?node2)\\n}\\nGROUP BY ?num_paralogy ?total ?result\",\n",
      "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
      "    \"question\": \"The percentage of proteins in Drosophila melanogaster that has at least one paralogous sequence (protein)\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nPREFIX ensemblterms: <http://rdf.ebi.ac.uk/terms/ensembl/>\\nSELECT DISTINCT ?gene_uniprot_uri ?ortholog_uniprot_uri {\\n\\tVALUES (?gene_uniprot_uri){\\n\\t\\t(<http://purl.uniprot.org/uniprot/P68871>)\\n\\t}\\n\\t?gene sio:SIO_000558 ?ortholog . # \\u00ab is orthologous to \\u00bb\\n\\t?gene obo:RO_0002162 ?taxon . # \\u00ab in taxon \\u00bb\\n\\t?ortholog obo:RO_0002162 ?ortholog_taxon.\\n\\t?ortholog ensemblterms:DEPENDENT ?ortholog_uniprot_uri.\\n\\t?gene ensemblterms:DEPENDENT ?gene_uniprot_uri.\\n\\tFILTER (\\n\\t\\t?taxon != ?ortholog_taxon\\n\\t)\\n}\",\n",
      "    \"endpoint\": \"https://www.ebi.ac.uk/rdf/services/sparql\",\n",
      "    \"question\": \"Retrieve from Ensembl dataset at the EBI RDF platform the pairwise orthologs of the gene with the P68871 UniProt accession number - only works if you disable CORS in your browser\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nPREFIX ensemblterms: <http://rdf.ebi.ac.uk/terms/ensembl/>\\nSELECT DISTINCT ?gene ?ortholog_uniprot_uri {\\n\\tVALUES (?gene){\\n\\t\\t(ensembl:ENSG00000139618)\\n\\t}\\n\\t?gene sio:SIO_000558 ?ortholog.\\n\\t?gene obo:RO_0002162 ?taxon.\\n\\t?ortholog obo:RO_0002162 ?ortholog_taxon.\\n\\t?ortholog ensemblterms:DEPENDENT ?ortholog_uniprot_uri.\\n\\t?gene ensemblterms:DEPENDENT ?gene_uniprot_uri.\\n\\tFILTER(\\n\\t\\t?taxon != ?ortholog_taxon\\n\\t\\t&&\\n\\t\\tSTRSTARTS(STR(?ortholog_uniprot_uri),\\\"http://purl.uniprot.org/uniprot/\\\")\\n\\t)\\n}\",\n",
      "    \"endpoint\": \"https://www.ebi.ac.uk/rdf/services/sparql\",\n",
      "    \"question\": \"Retrieve from Ensembl dataset at the EBI RDF platform the pairwise orthologs of the gene with the ENSG00000139618 Ensembl acession number - only works if you disable CORS in your browser\",\n",
      "    \"results\": 0\n",
      "  },\n",
      "  {\n",
      "    \"query\": \"PREFIX orthodb: <http://purl.orthodb.org/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT DISTINCT ?groups ?species_name ?protein1_uniprot ?gene1 ?taxLevel_uniprot ?taxLevel WHERE {\\n\\tVALUES ?protein2_uniprot {\\n\\t\\t<http://purl.uniprot.org/uniprot/P68871>\\n\\t}\\n\\tVALUES ?taxLevel {\\n\\t\\t\\\"Primates\\\"\\n\\t}\\n\\t?gene2 a orthodb:Gene.\\n\\t?gene2 orthodb:memberOf ?groups.\\n\\t?gene1 a orthodb:Gene.\\n\\t?gene1 orthodb:memberOf ?groups.\\n\\t?gene1 up:organism ?organism.\\n\\t?organism a ?taxon.\\n\\t?taxon up:scientificName ?species_name.\\n\\t?groups orthodb:ogBuiltAt ?taxLevel_uniprot.\\n\\t?taxLevel_uniprot up:scientificName ?taxLevel.\\n\\t?gene2 orthodb:xref ?xref2.\\n\\t?xref2 orthodb:xrefResource ?protein2_uniprot.\\n\\t?protein2_uniprot a orthodb:Uniprot.\\n\\t?gene1 orthodb:xref ?xref.\\n\\t?xref a orthodb:Xref.\\n\\tOPTIONAL {\\n\\t\\t?xref orthodb:xrefResource ?protein1_uniprot.\\n\\t\\t?protein1_uniprot a orthodb:Uniprot.\\n\\t}\\n}\\nORDER BY ?groups, ?taxLevel\",\n",
      "    \"endpoint\": \"https://sparql.orthodb.org\",\n",
      "    \"question\": \"Retrieve flat homologous groups (i.e. clusters) which are derived from the HOGs in the OrthoDB that contain the gene with the P68871 UniProt accession number\",\n",
      "    \"results\": 0\n",
      "  }\n",
      "]\n",
      "38 13\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "import requests\n",
    "from SPARQLWrapper import JSON, SPARQLWrapper\n",
    "\n",
    "queries_obj = requests.get(\n",
    "    \"https://raw.githubusercontent.com/biosoda/bioquery/master/biosoda_frontend/src/biosodadata.json\",\n",
    "    timeout=20,\n",
    ").json()\n",
    "\n",
    "# print(json.dumps(queries_obj, indent=2))\n",
    "\n",
    "get_datasources = {\n",
    "    \"species\": {\n",
    "        \"fetchUrl\": \"https://sparql.omabrowser.org/lode/servlet/query?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT DISTINCT ?value WHERE {\\n?s a up:Taxon .\\n?s up:scientificName ?value .\\n?s up:rank up:Species .\\nFILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n} LIMIT 100',\n",
    "        \"question\": \"Which species are available?\",\n",
    "    },\n",
    "    \"uniprotmnemonics\": {\n",
    "        \"fetchUrl\": \"https://sparql.uniprot.org/sparql?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?value\\nWHERE\\n{\\n ?protein a up:Protein .\\n ?protein up:mnemonic ?value\\nFILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n} LIMIT 100',\n",
    "        \"question\": \"What are the UniProt mnemonics matching the search string?\",\n",
    "    },\n",
    "    \"uniprottaxons\": {\n",
    "        \"fetchUrl\": \"https://sparql.uniprot.org/sparql?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT DISTINCT ?name ?value\\nWHERE\\n{\\n?value a up:Taxon ;\\nup:scientificName ?name\\nFILTER(CONTAINS(lcase(str(?name)), \"$$searchString$$\"))\\n} LIMIT 100',\n",
    "        \"question\": \"What are the taxonomic names and their corresponding IRIs?\",\n",
    "    },\n",
    "    \"uniprotnaturalannotations\": {\n",
    "        \"fetchUrl\": \"https://sparql.uniprot.org/sparql?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?value\\nWHERE\\n{\\n?protein a up:Protein .\\n?protein up:organism taxon:9606 .\\n?protein up:annotation ?annotation .\\n?annotation a up:Natural_Variant_Annotation .\\n?annotation rdfs:comment ?value .\\nFILTER (CONTAINS(?value, '$$searchString$$'))\\n} LIMIT 100\",\n",
    "        \"question\": \"What are the natural variant annotations for proteins in humans matching the search string?\",\n",
    "    },\n",
    "    \"ensemblgenes\": {\n",
    "        \"fetchUrl\": \"https://sparql.omabrowser.org/lode/servlet/query?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX dct: <http://purl.org/dc/terms/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT ?value\\nWHERE {\\n?protein1 sio:SIO_010079/lscr:xrefEnsemblGene ?ens .\\n\\n?ens dct:identifier ?value . FILTER(CONTAINS(lcase(str(?ens)), \"$$searchString$$\"))\\n} LIMIT 100',\n",
    "        \"question\": \"What are the Ensembl gene identifiers matching the search string?\",\n",
    "    },\n",
    "    \"omaproteins\": {\n",
    "        \"fetchUrl\": \"https://sparql.omabrowser.org/lode/servlet/query?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX dct: <http://purl.org/dc/terms/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT ?value\\nWHERE {\\n?protein1 a orth:Protein .\\n?protein1 dct:identifier ?value . FILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n} LIMIT 100',\n",
    "        \"question\": \"Which OMA protein identifiers match the search string?\",\n",
    "    },\n",
    "    \"taxlevels\": {\n",
    "        \"fetchUrl\": \"https://sparql.omabrowser.org/lode/servlet/query?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX orth: <http://purl.org/net/orth#>\\nSELECT DISTINCT ?value\\nWHERE {\\n?taxRange orth:taxRange ?value .\\nFILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n} LIMIT 100',\n",
    "        \"question\": \"What are the taxonomic levels matching the search string?\",\n",
    "    },\n",
    "    \"speciescommon\": {\n",
    "        \"fetchUrl\": \"https://sparql.uniprot.org/sparql?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX up: <http://purl.uniprot.org/core/> SELECT DISTINCT ?value WHERE {  ?s a up:Taxon .  ?s up:commonName ?value . ?s up:rank up:Species . FILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))}',\n",
    "        \"question\": \"What are the common names of species matching the search string?\",\n",
    "    },\n",
    "    \"speciescommonIRI\": {\n",
    "        \"fetchUrl\": \"https://sparql.uniprot.org/sparql?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX up: <http://purl.uniprot.org/core/> SELECT DISTINCT ?name ?value WHERE {  ?value a up:Taxon .  ?value up:commonName ?name . ?value up:rank up:Species . FILTER(CONTAINS(lcase(str(?name)), \"$$searchString$$\"))}',\n",
    "        \"question\": \"What are the common names and their corresponding IRIs of species matching the search string?\",\n",
    "    },\n",
    "    \"anatomicentities\": {\n",
    "        \"fetchUrl\": \"https://www.bgee.org/sparql/?query=$$query$$\",\n",
    "        \"queryHeaders\": {\"Accept\": \"application/sparql-results+json\"},\n",
    "        \"fetchQuery\": 'PREFIX genex: <http://purl.org/genex#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT DISTINCT ?value {\\n?anatEntity a genex:AnatomicalEntity .\\n?anatEntity rdfs:label ?value .\\nFILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n}',\n",
    "        \"question\": \"What are the anatomical entities matching the search string?\",\n",
    "    },\n",
    "    \"anatomicentities_extraspecies\": {\n",
    "        \"fetchUrl\": \"https://www.bgee.org/sparql/?query=$$query$$\",\n",
    "        \"queryHeaders\": {\"Accept\": \"application/sparql-results+json\"},\n",
    "        \"fetchQuery\": 'PREFIX genex: <http://purl.org/genex#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nSELECT DISTINCT ?value {\\n?gene genex:isExpressedIn ?anat .\\n?anat rdfs:label ?value .\\n?gene orth:organism ?o .\\n?o obo:RO_0002162 ?taxon .\\n?taxon up:commonName \"$$extra_0$$\" .\\n?anat a genex:AnatomicalEntity .\\nFILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n}',\n",
    "        \"question\": \"What are the anatomical entities for the species '$$extra_0$$' matching the search string?\",\n",
    "    },\n",
    "    \"stages\": {\n",
    "        \"fetchUrl\": \"https://www.bgee.org/sparql/?query=$$query$$\",\n",
    "        \"queryHeaders\": {\"Accept\": \"application/sparql-results+json\"},\n",
    "        \"fetchQuery\": 'PREFIX genex: <http://purl.org/genex#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT DISTINCT ?value { ?stage a <http://www.ebi.ac.uk/efo/EFO_0000399> .\\n?stage rdfs:label ?value .\\nFILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n} LIMIT 100',\n",
    "        \"question\": \"What are the developmental stages matching the search string?\",\n",
    "    },\n",
    "    \"genesbgee\": {\n",
    "        \"fetchUrl\": \"https://www.bgee.org/sparql/?query=$$query$$\",\n",
    "        \"queryHeaders\": {\"Accept\": \"application/sparql-results+json\"},\n",
    "        \"fetchQuery\": 'PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT DISTINCT ?value { ?gene a orth:Gene .\\n?gene rdfs:label ?value .\\nFILTER(STRSTARTS(lcase(str(?value)), \"$$searchString$$\"))\\n} LIMIT 100',\n",
    "        \"question\": \"What are the genes matching the search string?\",\n",
    "    },\n",
    "    \"speciesbgee\": {\n",
    "        \"fetchUrl\": \"https://www.bgee.org/sparql/?query=$$query$$\",\n",
    "        \"queryHeaders\": {\"Accept\": \"application/sparql-results+json\"},\n",
    "        \"fetchQuery\": 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT DISTINCT ?value {\\n?species a up:Taxon .\\n?species up:scientificName ?value .\\n?species up:rank up:Species .\\n\\nFILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n} ORDER BY ?value',\n",
    "        \"question\": \"What are the species names available in Bgee matching the search string?\",\n",
    "    },\n",
    "    \"speciesbgeecommon\": {\n",
    "        \"fetchUrl\": \"https://www.bgee.org/sparql/?query=$$query$$\",\n",
    "        \"queryHeaders\": {\"Accept\": \"application/sparql-results+json\"},\n",
    "        \"fetchQuery\": 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT DISTINCT ?value {\\n?species a up:Taxon .\\n?species up:commonName ?value .\\n?species up:rank up:Species .\\n\\nFILTER(CONTAINS(lcase(str(?value)), \"$$searchString$$\"))\\n} ORDER BY ?value',\n",
    "        \"question\": \"What are the common names of species available in Bgee matching the search string?\",\n",
    "    },\n",
    "    \"speciesbgeecommonIRI\": {\n",
    "        \"fetchUrl\": \"https://www.bgee.org/sparql/?query=$$query$$\",\n",
    "        \"queryHeaders\": {\"Accept\": \"application/sparql-results+json\"},\n",
    "        \"fetchQuery\": 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT DISTINCT ?name ?value {\\n?value a up:Taxon .\\n?value up:commonName ?name .\\n?value up:rank up:Species .\\n\\nFILTER(CONTAINS(lcase(str(?name)), \"$$searchString$$\"))\\n} ORDER BY ?value',\n",
    "        \"question\": \"What are the common names and their corresponding IRIs of species available in Bgee matching the search string?\",\n",
    "    },\n",
    "    \"proteinsatoma\": {\n",
    "        \"fetchUrl\": \"https://sparql.omabrowser.org/lode/servlet/query?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nSELECT DISTINCT ?value WHERE {\\n?proteinOMA a orth:Protein;\\nrdfs:label ?value .\\nFILTER(STRSTARTS(lcase(str(?value)), \"$$searchString$$\"))}',\n",
    "        \"question\": \"What are the protein identifiers at OMA matching the search string?\",\n",
    "    },\n",
    "    \"proteins\": {\n",
    "        \"fetchUrl\": \"https://sparql.uniprot.org/sparql?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nSELECT DISTINCT ?value\\nWHERE\\n{\\n?protein a up:Protein .\\n?protein up:recommendedName ?recommended .\\n?recommended up:fullName ?value .\\n?protein up:encodedBy ?gene .\\n?gene skos:prefLabel ?text .\\nFILTER CONTAINS(?text, '$$searchString$$')\\n} LIMIT 100\",\n",
    "        \"question\": \"What are the protein names matching the search string?\",\n",
    "    },\n",
    "    \"genesuniprot\": {\n",
    "        \"fetchUrl\": \"https://sparql.uniprot.org/sparql?query=$$query$$&format=JSON$$&limit=limit$$&offset=$$offset$$&inference=false\",\n",
    "        \"fetchQuery\": \"PREFIX up:<http://purl.uniprot.org/core/>\\nPREFIX taxon:<http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX skos:<http://www.w3.org/2004/02/skos/core#>\\nSELECT DISTINCT ?name ?value WHERE {\\n\\t?value a up:Protein . \\n\\t?value up:organism taxon:9606 . \\n\\t?value up:encodedBy ?gene . \\n\\t?gene skos:prefLabel ?name . \\n\\tFILTER CONTAINS(lcase(str(?name)), lcase('$$searchString$$'))\\n} ORDER BY ?value LIMIT 100\",\n",
    "        \"question\": \"What are the gene names and their corresponding IRIs for proteins in humans matching the search string?\",\n",
    "    },\n",
    "}\n",
    "\n",
    "\n",
    "def exec_query(query: str, endpoint: str) -> int:\n",
    "    try:\n",
    "        sparql_endpoint = SPARQLWrapper(endpoint)\n",
    "        sparql_endpoint.setReturnFormat(JSON)\n",
    "        sparql_endpoint.setTimeout(200)\n",
    "        sparql_endpoint.setQuery(query)\n",
    "        results = sparql_endpoint.query().convert()\n",
    "        return len(results[\"results\"][\"bindings\"])\n",
    "    except Exception as _e:\n",
    "        return 0\n",
    "\n",
    "\n",
    "# Process each question\n",
    "results = []\n",
    "failed = []\n",
    "for i, question in enumerate(queries_obj[\"questions\"]):\n",
    "    if \"SPARQL\" in question and \"vars\" in question:\n",
    "        sparql_query = question[\"SPARQL\"]\n",
    "        question_str = question[\"question\"]\n",
    "        for var in question[\"vars\"]:\n",
    "            sparql_query = sparql_query.replace(\n",
    "                f\"$${var['name']}$$\", var.get(\"defaultvalue\", var[\"default\"])\n",
    "            )\n",
    "            question_str = question_str.replace(f\"$${var['name']}$$\", var[\"default\"])\n",
    "\n",
    "        print(f\"[{i}/{len(queries_obj['questions'])}] {question_str}\")\n",
    "        sparql_query = sparql_query.replace(\"$$innerlimit$$\", \"LIMIT 10\")\n",
    "        # sparql_query = sparql_query.replace(\"$$innerlimit$$\", \"\")\n",
    "\n",
    "        endpoint_url = (\n",
    "            \"https://www.bgee.org/sparql/\"\n",
    "            if question[\"fetchUrlShort\"].startswith(\"https://biosoda.expasy.org:4443\")\n",
    "            else question[\"fetchUrlShort\"].replace(\n",
    "                \"https://sparql.omabrowser.org/lode/sparql\",\n",
    "                \"https://sparql.omabrowser.org/sparql/\",\n",
    "            )\n",
    "        )\n",
    "        res_count = exec_query(sparql_query, endpoint_url)\n",
    "        # res_count = 1\n",
    "        if res_count > 0:\n",
    "            print(endpoint_url)\n",
    "            print(sparql_query)\n",
    "            results.append(\n",
    "                {\n",
    "                    \"query\": sparql_query,\n",
    "                    \"endpoint\": endpoint_url,\n",
    "                    \"question\": question_str,\n",
    "                    \"results\": res_count,\n",
    "                }\n",
    "            )\n",
    "        else:\n",
    "            failed.append(\n",
    "                {\n",
    "                    \"query\": sparql_query,\n",
    "                    \"endpoint\": endpoint_url,\n",
    "                    \"question\": question_str,\n",
    "                    \"results\": res_count,\n",
    "                }\n",
    "            )\n",
    "        print(f\"{len(results)} / {len(failed)}\")\n",
    "\n",
    "# Resolve entities using datasources?\n",
    "# https://github.com/biosoda/bioquery/blob/master/biosoda_frontend/src/biosodadata.json#L1491\n",
    "\n",
    "# Output the results\n",
    "print(json.dumps(results, indent=2))\n",
    "print(json.dumps(failed, indent=2))\n",
    "print(len(results), len(failed))\n",
    "\n",
    "# NOTE: all process up to 13/52"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-scripts",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
