{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running example queries\n",
    "\n",
    "Trying to run example SPARQL queries using `SPARQLWrapper`. Painful experience, the wrapper is not working well even with basic Virtuoso SPARQL endpoint\n",
    "\n",
    "## Questions to test\n",
    "\n",
    "Select all human UniProt entries with a sequence variant that leads to a tyrosine to phenylalanine substitution\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "129\n",
      "[0/129] Was any UniProt entry integrated on the 9th of January 2013 <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 1 in 0.09s\n",
      "[1/129] Construct new triples of the type 'HumanProtein' from all human UniProt entries <https://sparql.uniprot.org/sparql/>\n",
      "[2/129] Select all taxa from the UniProt taxonomy <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 2941742 in 321.27s\n",
      "[3/129] Select all human UniProt entries with a sequence variant that leads to a tyrosine to phenylalanine substitution <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 128 in 9.82s\n",
      "[4/129] Select all UniProt entries with annotated transmembrane regions and the regions' begin and end coordinates on the canonical sequence <https://sparql.uniprot.org/sparql/>\n",
      "üí£Ô∏è results The read operation timed out in 5294.79s\n",
      "[5/129] Select all UniProt entries that were integrated on the 30th of November 2010 <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 643989 in 180.40s\n",
      "[6/129] Select all triples that relate to the EMBL CDS entry AA089367.1: <https://sparql.uniprot.org/sparql/>\n",
      "[7/129] Select all triples that relate to the taxon that describes Homo sapiens in the named graph for taxonomy <https://sparql.uniprot.org/sparql/>\n",
      "[8/129] Select the average number of cross-references to the PDB database of UniProt entries that have at least one cross-reference to the PDB database <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 1 in 52.22s\n",
      "[9/129] Select the number of UniProt entries for each of the EC (Enzyme Commission) top level categories <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 7 in 1.86s\n",
      "[10/129] Find all Natural Variant Annotations if associated via an evidence tag to an article with a pubmed identifier <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 197124 in 49.71s\n",
      "[11/129] Select all bacterial taxa and their scientific name from the UniProt taxonomy <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 556591 in 62.38s\n",
      "[12/129] Find how often an article in pubmed was used in an evidence tag in a human protein (ordered by most used to least) <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 43889 in 53.30s\n",
      "[13/129] Find where disease related proteins are known to be located in the cell <https://sparql.uniprot.org/sparql/>\n",
      "üí£Ô∏è results The read operation timed out in 250.71s\n",
      "[14/129] For two accessions find the GO term labels and group them into GO process,function and component <https://sparql.uniprot.org/sparql/>\n",
      "[\n",
      "    {\n",
      "        \"question\": \"Was any UniProt entry integrated on the 9th of January 2013\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nASK\\nWHERE\\n{\\n\\t?protein a up:Protein .\\n\\t?protein up:created '2013-01-09'^^xsd:date\\n}\\n\",\n",
      "        \"results\": 1,\n",
      "        \"runtime\": 0\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all taxa from the UniProt taxonomy\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?taxon\\nFROM <http://sparql.uniprot.org/taxonomy>\\nWHERE\\n{\\n    ?taxon a up:Taxon .\\n}\\n\",\n",
      "        \"results\": 2941742,\n",
      "        \"runtime\": 321\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all human UniProt entries with a sequence variant that leads to a tyrosine to phenylalanine substitution\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT ?protein ?annotation ?begin ?text\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:organism taxon:9606 ; \\n            up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation ;\\n            rdfs:comment ?text ;\\n            up:substitution ?substitution ;\\n            up:range/faldo:begin\\n                [ faldo:position ?begin ;\\n                  faldo:reference ?sequence ] .\\n        ?sequence rdf:value ?value .\\n        BIND (substr(?value, ?begin, 1) as ?original) .\\n        FILTER(?original = 'Y' && ?substitution = 'F') .\\n} \\n\",\n",
      "        \"results\": 128,\n",
      "        \"runtime\": 9\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all UniProt entries that were integrated on the 30th of November 2010\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein\\nWHERE\\n{\\n\\t?protein a up:Protein . \\n\\t?protein up:created '2010-11-30'^^xsd:date\\n} \\n\",\n",
      "        \"results\": 643989,\n",
      "        \"runtime\": 180\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select the average number of cross-references to the PDB database of UniProt entries that have at least one cross-reference to the PDB database\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT (AVG(?linksToPdbPerEntry) AS ?avgLinksToPdbPerEntry)\\nWHERE\\n{\\n\\tSELECT ?protein (COUNT(DISTINCT ?db) AS ?linksToPdbPerEntry)\\n\\tWHERE\\n\\t{\\n\\t\\t?protein a up:Protein .\\n\\t\\t?protein rdfs:seeAlso ?db .\\n\\t\\t?db up:database <http://purl.uniprot.org/database/PDB> .\\n\\t}\\n\\tGROUP BY ?protein ORDER BY DESC(?linksToPdbPerEntry)\\n}\\n\",\n",
      "        \"results\": 1,\n",
      "        \"runtime\": 52\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select the number of UniProt entries for each of the EC (Enzyme Commission) top level categories\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ec: <http://purl.uniprot.org/enzyme/>\\nSELECT ?ecClass (COUNT(?protein) as ?size)\\nWHERE\\n{\\n    VALUES (?ecClass) {(ec:1.-.-.-) (ec:2.-.-.-) (ec:3.-.-.-) (ec:4.-.-.-) (ec:5.-.-.-) (ec:6.-.-.-) (ec:7.-.-.-)} .\\n    ?protein ( up:enzyme | up:domain/up:enzyme | up:component/up:enzyme ) ?enzyme .\\n    # Enzyme subclasses are materialized, do not use rdfs:subClassOf+\\n    ?enzyme rdfs:subClassOf ?ecClass .\\n}\\nGROUP BY ?ecClass ORDER BY ?ecClass\\n\\n\",\n",
      "        \"results\": 7,\n",
      "        \"runtime\": 1\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find all Natural Variant Annotations if associated via an evidence tag to an article with a pubmed identifier\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?accession\\n    ?annotation_acc \\n    ?pubmed\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation .\\n        ?linkToEvidence rdf:object ?annotation ;\\n                        up:attribution ?attribution .\\n        ?attribution up:source ?source .\\n        ?source a up:Journal_Citation .\\n  BIND(SUBSTR(STR(?protein),33) AS ?accession)\\n  BIND(IF(CONTAINS(STR(?annotation), \\\"#SIP\\\"), SUBSTR(STR(?annotation),33), SUBSTR(STR(?annotation),36))AS?annotation_acc)\\n  BIND(SUBSTR(STR(?source),35) AS ?pubmed)\\n}\\n\\n\",\n",
      "        \"results\": 197124,\n",
      "        \"runtime\": 49\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all bacterial taxa and their scientific name from the UniProt taxonomy\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?taxon ?name\\nWHERE\\n{\\n    ?taxon a up:Taxon .\\n    ?taxon up:scientificName ?name .\\n    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\\n    ?taxon rdfs:subClassOf taxon:2 .\\n}\\n\",\n",
      "        \"results\": 556591,\n",
      "        \"runtime\": 62\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find how often an article in pubmed was used in an evidence tag in a human protein (ordered by most used to least)\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?source \\n    (COUNT(?attribution) AS ?attribitions)\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:organism taxon:9606 ;\\n            up:annotation ?annotation .\\n        ?linkToEvidence rdf:object ?annotation ;\\n                        up:attribution ?attribution .\\n        ?attribution up:source ?source .\\n        ?source a up:Journal_Citation .\\n} GROUP BY ?source ORDER BY DESC(COUNT(?attribution))\\n\\n\",\n",
      "        \"results\": 43889,\n",
      "        \"runtime\": 53\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"For two accessions find the GO term labels and group them into GO process,function and component\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n\\nSELECT \\n    (CONCAT(SUBSTR(STR(?protein), 33)) AS ?uniprot)\\n    (GROUP_CONCAT(?celtype; separator=\\\";\\\") AS ?celtypes)\\n    (GROUP_CONCAT(?biotype; separator=\\\";\\\") AS ?biotypes)\\n    (GROUP_CONCAT(?moltype; separator=\\\";\\\") AS ?moltypes)\\nWHERE\\n{\\n    VALUES (?ac) {(\\\"Q6GZX4\\\") (\\\"Q96375\\\")}\\n    BIND (IRI(CONCAT(\\\"http://purl.uniprot.org/uniprot/\\\",?ac)) AS ?protein)\\n    ?protein a up:Protein .\\n    ?protein up:classifiedWith ?goTerm .\\n    #Determine if the type is biological_process\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0008150>.\\n        ?goTerm rdfs:label ?biotype .\\n    }\\n    #Determine if the type is cellular_component\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0005575>.\\n        ?goTerm rdfs:label ?celtype .\\n    }\\n    #Determine if the type is molecular_function\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0003674> .\\n        ?goTerm rdfs:label ?moltype .\\n    }\\n    #Filter out the uniprot keywords\\n    FILTER(bound(?biotype) || bound(?celtype) || bound(?moltype))\\n} GROUP BY ?protein\\n\\n\",\n",
      "        \"results\": 2,\n",
      "        \"runtime\": 0\n",
      "    }\n",
      "]\n",
      "‚úÖ results 2 in 0.25s\n",
      "[15/129] Number of reviewed entries (Swiss-Prot) that are related to kinase activity <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 1 in 23.63s\n",
      "[16/129] Find the release number of the uniprot data that is currently being queried <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 1 in 0.04s\n",
      "[17/129] Find any uniprot entry which has a name 'HLA class I histocompatibility antigen, B-73 alpha chain' <https://sparql.uniprot.org/sparql/>\n",
      "‚ùå results 0 in 0.16s\n",
      "[18/129] Find any uniprot entry, or an uniprot entries domain or component which has a name 'HLA class I histocompatibility antigen, B-73 alpha chain' <https://sparql.uniprot.org/sparql/>\n",
      "‚ùå results 0 in 0.08s\n",
      "[19/129] Find all names associated with uniprot entry P05067, and if the name is associated with the entry it's domains or its components <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 50 in 115.50s\n",
      "[20/129] Get the list of uniprot entries for the chromosome of proteome UP000000625 <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 8802 in 3.24s\n",
      "[21/129] Use ALLIE a service for Abbreviation / Long Form in Japanese and English to search in UniProt using Japanese. <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 51 in 2.15s\n",
      "[22/129] Select all UniProt entries, and their organism and amino acid sequences (including isoforms), for E. coli K12 and all its strains <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 3746 in 4.83s\n",
      "[23/129] Find UniProt entries with merged loci in Bordetella avium <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 3349 in 11.60s\n",
      "[24/129] Find UniParc records whose sequence point to the most database entries <https://sparql.uniprot.org/sparql/>\n",
      "üí£Ô∏è results The read operation timed out in 200.13s\n",
      "[25/129] Find UniProtKB entries with more than 1 Topological domain annotation <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 175067 in 32.82s\n",
      "[26/129] Find longest comment text associated with a UniProtKB Natural Variant Annotation <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 66010 in 8.22s\n",
      "[27/129] Find the co-occurence count of Topological Domain comment text in UniProtKB entries <https://sparql.uniprot.org/sparql/>\n",
      "[\n",
      "    {\n",
      "        \"question\": \"Was any UniProt entry integrated on the 9th of January 2013\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nASK\\nWHERE\\n{\\n\\t?protein a up:Protein .\\n\\t?protein up:created '2013-01-09'^^xsd:date\\n}\\n\",\n",
      "        \"results\": 1,\n",
      "        \"runtime\": 0\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all taxa from the UniProt taxonomy\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?taxon\\nFROM <http://sparql.uniprot.org/taxonomy>\\nWHERE\\n{\\n    ?taxon a up:Taxon .\\n}\\n\",\n",
      "        \"results\": 2941742,\n",
      "        \"runtime\": 321\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all human UniProt entries with a sequence variant that leads to a tyrosine to phenylalanine substitution\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT ?protein ?annotation ?begin ?text\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:organism taxon:9606 ; \\n            up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation ;\\n            rdfs:comment ?text ;\\n            up:substitution ?substitution ;\\n            up:range/faldo:begin\\n                [ faldo:position ?begin ;\\n                  faldo:reference ?sequence ] .\\n        ?sequence rdf:value ?value .\\n        BIND (substr(?value, ?begin, 1) as ?original) .\\n        FILTER(?original = 'Y' && ?substitution = 'F') .\\n} \\n\",\n",
      "        \"results\": 128,\n",
      "        \"runtime\": 9\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all UniProt entries that were integrated on the 30th of November 2010\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein\\nWHERE\\n{\\n\\t?protein a up:Protein . \\n\\t?protein up:created '2010-11-30'^^xsd:date\\n} \\n\",\n",
      "        \"results\": 643989,\n",
      "        \"runtime\": 180\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select the average number of cross-references to the PDB database of UniProt entries that have at least one cross-reference to the PDB database\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT (AVG(?linksToPdbPerEntry) AS ?avgLinksToPdbPerEntry)\\nWHERE\\n{\\n\\tSELECT ?protein (COUNT(DISTINCT ?db) AS ?linksToPdbPerEntry)\\n\\tWHERE\\n\\t{\\n\\t\\t?protein a up:Protein .\\n\\t\\t?protein rdfs:seeAlso ?db .\\n\\t\\t?db up:database <http://purl.uniprot.org/database/PDB> .\\n\\t}\\n\\tGROUP BY ?protein ORDER BY DESC(?linksToPdbPerEntry)\\n}\\n\",\n",
      "        \"results\": 1,\n",
      "        \"runtime\": 52\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select the number of UniProt entries for each of the EC (Enzyme Commission) top level categories\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ec: <http://purl.uniprot.org/enzyme/>\\nSELECT ?ecClass (COUNT(?protein) as ?size)\\nWHERE\\n{\\n    VALUES (?ecClass) {(ec:1.-.-.-) (ec:2.-.-.-) (ec:3.-.-.-) (ec:4.-.-.-) (ec:5.-.-.-) (ec:6.-.-.-) (ec:7.-.-.-)} .\\n    ?protein ( up:enzyme | up:domain/up:enzyme | up:component/up:enzyme ) ?enzyme .\\n    # Enzyme subclasses are materialized, do not use rdfs:subClassOf+\\n    ?enzyme rdfs:subClassOf ?ecClass .\\n}\\nGROUP BY ?ecClass ORDER BY ?ecClass\\n\\n\",\n",
      "        \"results\": 7,\n",
      "        \"runtime\": 1\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find all Natural Variant Annotations if associated via an evidence tag to an article with a pubmed identifier\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?accession\\n    ?annotation_acc \\n    ?pubmed\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation .\\n        ?linkToEvidence rdf:object ?annotation ;\\n                        up:attribution ?attribution .\\n        ?attribution up:source ?source .\\n        ?source a up:Journal_Citation .\\n  BIND(SUBSTR(STR(?protein),33) AS ?accession)\\n  BIND(IF(CONTAINS(STR(?annotation), \\\"#SIP\\\"), SUBSTR(STR(?annotation),33), SUBSTR(STR(?annotation),36))AS?annotation_acc)\\n  BIND(SUBSTR(STR(?source),35) AS ?pubmed)\\n}\\n\\n\",\n",
      "        \"results\": 197124,\n",
      "        \"runtime\": 49\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all bacterial taxa and their scientific name from the UniProt taxonomy\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?taxon ?name\\nWHERE\\n{\\n    ?taxon a up:Taxon .\\n    ?taxon up:scientificName ?name .\\n    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\\n    ?taxon rdfs:subClassOf taxon:2 .\\n}\\n\",\n",
      "        \"results\": 556591,\n",
      "        \"runtime\": 62\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find how often an article in pubmed was used in an evidence tag in a human protein (ordered by most used to least)\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?source \\n    (COUNT(?attribution) AS ?attribitions)\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:organism taxon:9606 ;\\n            up:annotation ?annotation .\\n        ?linkToEvidence rdf:object ?annotation ;\\n                        up:attribution ?attribution .\\n        ?attribution up:source ?source .\\n        ?source a up:Journal_Citation .\\n} GROUP BY ?source ORDER BY DESC(COUNT(?attribution))\\n\\n\",\n",
      "        \"results\": 43889,\n",
      "        \"runtime\": 53\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"For two accessions find the GO term labels and group them into GO process,function and component\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n\\nSELECT \\n    (CONCAT(SUBSTR(STR(?protein), 33)) AS ?uniprot)\\n    (GROUP_CONCAT(?celtype; separator=\\\";\\\") AS ?celtypes)\\n    (GROUP_CONCAT(?biotype; separator=\\\";\\\") AS ?biotypes)\\n    (GROUP_CONCAT(?moltype; separator=\\\";\\\") AS ?moltypes)\\nWHERE\\n{\\n    VALUES (?ac) {(\\\"Q6GZX4\\\") (\\\"Q96375\\\")}\\n    BIND (IRI(CONCAT(\\\"http://purl.uniprot.org/uniprot/\\\",?ac)) AS ?protein)\\n    ?protein a up:Protein .\\n    ?protein up:classifiedWith ?goTerm .\\n    #Determine if the type is biological_process\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0008150>.\\n        ?goTerm rdfs:label ?biotype .\\n    }\\n    #Determine if the type is cellular_component\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0005575>.\\n        ?goTerm rdfs:label ?celtype .\\n    }\\n    #Determine if the type is molecular_function\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0003674> .\\n        ?goTerm rdfs:label ?moltype .\\n    }\\n    #Filter out the uniprot keywords\\n    FILTER(bound(?biotype) || bound(?celtype) || bound(?moltype))\\n} GROUP BY ?protein\\n\\n\",\n",
      "        \"results\": 2,\n",
      "        \"runtime\": 0\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Number of reviewed entries (Swiss-Prot) that are related to kinase activity\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX GO: <http://purl.obolibrary.org/obo/GO_>\\nSELECT\\n    (COUNT(DISTINCT(?protein)) AS ?pc)\\nWHERE\\n{   \\n    ?protein rdf:type up:Protein ;\\n        up:reviewed true  ;\\n        up:organism taxon:9606 ;\\n        up:classifiedWith|(up:classifiedWith/rdfs:subClassOf) GO:0016301 .\\n}\\n\",\n",
      "        \"results\": 1,\n",
      "        \"runtime\": 23\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find the release number of the uniprot data that is currently being queried\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nSELECT ?version\\nFROM <https://sparql.uniprot.org/.well-known/void>\\nWHERE\\n{\\n    [] <http://purl.org/pav/version> ?version\\n}\\n\",\n",
      "        \"results\": 1,\n",
      "        \"runtime\": 0\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find all names associated with uniprot entry P05067, and if the name is associated with the entry it's domains or its components\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?protein\\n  ?anyKindOfName \\n  ?names \\n  ?partType\\nWHERE\\n{\\n  BIND(<http://purl.uniprot.org/uniprot/P05067> AS ?protein)\\n  ?protein a up:Protein .\\n  {\\n    ?protein (up:recommendedName|up:alternativeName) ?structuredName .\\n  }\\n    UNION\\n  {\\n    VALUES(?partType){(up:domain) (up:component)}\\n    ?protein ?partType ?part .\\n    ?part (up:recommendedName|up:alternativeName) ?structuredName .\\n  }\\n  ?structuredName ?anyKindOfName  ?names .\\n  ?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\\n}\\n\",\n",
      "        \"results\": 50,\n",
      "        \"runtime\": 115\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Get the list of uniprot entries for the chromosome of proteome UP000000625\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nSELECT \\n  ?protein\\n  ?proteome \\nWHERE\\n{\\n  ?protein a up:Protein ;\\n           up:reviewed true ;\\n           up:proteome ?proteome .\\n  VALUES (?proteome) {(<http://purl.uniprot.org/proteomes/UP000000625#Chromosome>)}\\n}\\n\",\n",
      "        \"results\": 8802,\n",
      "        \"runtime\": 3\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Use ALLIE a service for Abbreviation / Long Form in Japanese and English to search in UniProt using Japanese.\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n    ?protein ?englishLabelStr\\nWHERE {\\n    SERVICE<http://data.allie.dbcls.jp/sparql>{\\n            ?x rdfs:label \\\"\\u30a2\\u30df\\u30ed\\u30a4\\u30c9\\u524d\\u99c6\\u4f53\\u30bf\\u30f3\\u30d1\\u30af\\u8cea\\\"@ja ;\\n                rdfs:label ?englishLabel .\\n            FILTER(lang(?englishLabel) = \\\"en\\\")\\n    }\\n    BIND (STR(?englishLabel) AS ?englishLabelStr)\\n    ?protein a up:Protein .\\n    {\\n        ?protein (up:recommendedName|up:alternativeName) ?structuredName .\\n    }\\n    UNION\\n    {\\n          VALUES(?partType){(up:domain) (up:component)}\\n                  ?protein ?partType ?part .\\n          ?part (up:recommendedName|up:alternativeName) ?structuredName .\\n    }\\n    ?structuredName ?anyKindOfName  ?englishLabelStr .\\n    ?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\\n}\\n\\n\",\n",
      "        \"results\": 51,\n",
      "        \"runtime\": 2\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Select all UniProt entries, and their organism and amino acid sequences (including isoforms), for E. coli K12 and all its strains\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT ?protein ?organism ?isoform ?sequence\\nWHERE\\n{\\n    ?protein a up:Protein .\\n    ?protein up:organism ?organism .\\n    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\\n    ?organism rdfs:subClassOf taxon:83333 .\\n    ?protein up:sequence ?isoform .\\n    ?isoform rdf:value ?sequence .\\n}\\n\",\n",
      "        \"results\": 3746,\n",
      "        \"runtime\": 4\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find UniProt entries with merged loci in Bordetella avium\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nSELECT\\n  ?protein \\n  (GROUP_CONCAT(?locusName; separator=',') AS ?locusNames)\\nWHERE \\n{ \\n  ?protein a up:Protein ;\\n    up:organism taxon:360910 ;\\n    up:encodedBy ?gene .\\n  ?gene up:locusName ?locusName .\\n} \\nGROUP BY ?protein \\nHAVING (COUNT(?locusName) > 1)\\n\\n\",\n",
      "        \"results\": 3349,\n",
      "        \"runtime\": 11\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find UniProtKB entries with more than 1 Topological domain annotation\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?protein \\n    (GROUP_CONCAT(?comment; separator=\\\", \\\") AS ?comments)\\nWHERE\\n{\\n    ?protein a up:Protein ;\\n            up:annotation ?annotation . \\n    ?annotation rdf:type up:Topological_Domain_Annotation ;\\n            rdfs:comment ?comment .\\n} \\nGROUP BY ?protein \\nHAVING (COUNT(?annotation) > 1)\\n\\n\",\n",
      "        \"results\": 175067,\n",
      "        \"runtime\": 32\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find longest comment text associated with a UniProtKB Natural Variant Annotation\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT \\n    ?annotation ?comment\\nWHERE {\\n    ?annotation a up:Natural_Variant_Annotation ;\\n        rdfs:comment ?comment . \\n} \\nORDER BY DESC(STRLEN(?comment))\\n\\n\",\n",
      "        \"results\": 66010,\n",
      "        \"runtime\": 8\n",
      "    },\n",
      "    {\n",
      "        \"question\": \"Find the co-occurence count of Topological Domain comment text in UniProtKB entries\",\n",
      "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
      "        \"query\": \"\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?comment1 \\n    ?comment2 \\n    (COUNT(?comment1) AS ?count1)\\nWHERE\\n{\\n    ?protein a up:Protein ;\\n               up:annotation ?annotation1 , \\n                             ?annotation2 . \\n    ?annotation1 rdf:type up:Topological_Domain_Annotation ;\\n        rdfs:comment ?rawComment1 .\\n    ?annotation2 rdf:type up:Topological_Domain_Annotation ;\\n        rdfs:comment ?rawComment2 . \\n    BIND(IF(contains(?rawComment1, ';'), \\n            STRBEFORE(?rawComment1,';'), \\n            ?rawComment1) AS ?comment1)\\n    BIND(IF(contains(?rawComment2, ';'), \\n            STRBEFORE(?rawComment2,';'), \\n            ?rawComment2) AS ?comment2)\\n    FILTER(?annotation1 != ?annotation2)\\n} \\nGROUP BY ?comment1 ?comment2 \\nORDER BY DESC(COUNT(?comment1))\\n\\n\",\n",
      "        \"results\": 118,\n",
      "        \"runtime\": 22\n",
      "    }\n",
      "]\n",
      "‚úÖ results 118 in 22.12s\n",
      "[28/129] Find the similar proteins for UniProtKB entry P05067 sorted by UniRef cluster identity <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 79 in 0.13s\n",
      "[29/129] Find the orthologous proteins for UniProtKB entry P05067 using the OrthoDB database <https://sparql.uniprot.org/sparql/>\n",
      "üí£Ô∏è results The read operation timed out in 200.14s\n",
      "[30/129] Find the human protein which contains an Epitope VSTQ, where T is a phosporylated threonine <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 2 in 46.69s\n",
      "[31/129] For the human entry P05067 (Amyloid-beta precursor protein) find the gene start ends in WikiData <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 2 in 0.75s\n",
      "[32/129] Retrieve entries and Catalytic activities in the reviewed (Swiss-Prot) section that have experimental evidences, <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 47936 in 26.74s\n",
      "[33/129] Select the UniProt entry with the mnemonic 'A4_HUMAN' <https://sparql.uniprot.org/sparql/>\n",
      "‚úÖ results 1 in 0.06s\n",
      "[34/129] Retrieve human enzymes that metabolize sphingolipids and are annotated in ChEMBL <https://sparql.uniprot.org/sparql/>\n",
      "üí£Ô∏è results The read operation timed out in 200.07s\n",
      "[35/129] Retrieve entries with sequences that are composed of fragments <https://sparql.uniprot.org/sparql/>\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import re\n",
    "import time\n",
    "\n",
    "from qdrant_client.models import FieldCondition, Filter, MatchValue\n",
    "from rdflib.plugins.sparql import prepareQuery\n",
    "from SPARQLWrapper import JSON, TURTLE, SPARQLWrapper\n",
    "\n",
    "from sparql_llm.config import settings\n",
    "from sparql_llm.index import get_vectordb\n",
    "\n",
    "vectordb = get_vectordb(\"localhost\")\n",
    "all_queries, _ = vectordb.scroll(\n",
    "    collection_name=settings.docs_collection_name,\n",
    "    scroll_filter=Filter(\n",
    "        must=[\n",
    "            FieldCondition(\n",
    "                key=\"doc_type\",\n",
    "                match=MatchValue(value=\"sparql\"),\n",
    "            )\n",
    "        ]\n",
    "    ),\n",
    "    limit=1000,\n",
    "    with_payload=True,\n",
    "    with_vectors=False,\n",
    ")\n",
    "\n",
    "\n",
    "def ensure_limit(query: str) -> str:\n",
    "    limit_pattern = re.compile(r\"LIMIT\\s+\\d+\\s*$\", re.IGNORECASE)\n",
    "    if not limit_pattern.search(query.rstrip()):\n",
    "        query = query.rstrip() + \" LIMIT 1\"\n",
    "    return query\n",
    "\n",
    "\n",
    "questions_results = []\n",
    "\n",
    "queries_returning_rdf = [\"DescribeQuery\", \"ConstructQuery\"]\n",
    "# skip_queries = [7, 27, 32, 37, 38, 42, 43, 47, 62, 63]\n",
    "print(len(all_queries))\n",
    "for i, query_record in enumerate(all_queries):\n",
    "    # if i < 63:\n",
    "    #     continue\n",
    "    start_time = time.time()\n",
    "    try:\n",
    "        # print(query_record.payload)\n",
    "        question = query_record.payload[\"question\"].split(\":\", 1)[1].strip()\n",
    "        endpoint = query_record.payload[\"endpoint\"]\n",
    "        query = (\n",
    "            query_record.payload[\"answer\"].removeprefix(\"```sparql\").removesuffix(\"```\")\n",
    "        )\n",
    "        # if i in skip_queries:\n",
    "        #     print(f\"‚è©Ô∏è {i}/{len(all_queries)} Skipping to avoid timeout: {question} <{endpoint}>\")\n",
    "        #     continue\n",
    "\n",
    "        print(f\"[{i}/{len(all_queries)}] {question} <{endpoint}>\")\n",
    "\n",
    "        parsed_query = prepareQuery(query)\n",
    "        # print(parsed_query.algebra.name)\n",
    "\n",
    "        sparql_endpoint = SPARQLWrapper(endpoint)\n",
    "        if parsed_query.algebra.name not in queries_returning_rdf:\n",
    "            sparql_endpoint.setReturnFormat(JSON)\n",
    "        else:\n",
    "            sparql_endpoint.setReturnFormat(TURTLE)\n",
    "            continue\n",
    "\n",
    "        sparql_endpoint.setQuery(query)\n",
    "        sparql_endpoint.setTimeout(200)\n",
    "\n",
    "        # NOTE: for some reason CONSTRUCT queries are failing directly with every format\n",
    "        results = sparql_endpoint.query().convert()\n",
    "        # results = sparql_endpoint.queryAndConvert()\n",
    "        if parsed_query.algebra.name in queries_returning_rdf:\n",
    "            # print(results)\n",
    "            res_count = len(results)\n",
    "        elif parsed_query.algebra.name == \"AskQuery\":\n",
    "            res_count = 1 if results[\"boolean\"] else 0\n",
    "        else:\n",
    "            res_count = len(results[\"results\"][\"bindings\"])\n",
    "\n",
    "        string_resp = \"‚úÖ\" if res_count > 0 else \"‚ö†Ô∏è\"\n",
    "        if res_count > 0:\n",
    "            string_resp = \"‚úÖ\"\n",
    "            questions_results.append(\n",
    "                {\n",
    "                    \"question\": question,\n",
    "                    \"endpoint\": endpoint,\n",
    "                    \"query\": query,\n",
    "                    \"results\": res_count,\n",
    "                    \"runtime\": int(time.time() - start_time),\n",
    "                }\n",
    "            )\n",
    "        else:\n",
    "            string_resp = \"‚ùå\"\n",
    "        if len(questions_results) % 10 == 0:\n",
    "            print(json.dumps(questions_results, indent=4))\n",
    "    except Exception as e:\n",
    "        res_count = str(e)\n",
    "        string_resp = \"üí£Ô∏è\"\n",
    "        # print(query)\n",
    "\n",
    "    print(f\"{string_resp} results {res_count} in {time.time() - start_time:.2f}s\")\n",
    "\n",
    "print(json.dumps(questions_results, indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-scripts",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
