{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing a SPARQL query to extract basic graph patterns\n",
    "\n",
    "Function to parse a SPARQL query and extract the BGP it contains to load them in a RDFLib graph, with support for federated queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subject ?hgnc in endpoint https://sparql.uniprot.org/sparql does not support the predicate rdfs:label. Correct predicate might be one of the following: owl:sameAs, rdfs:seeAlso, up:transcribedFrom, up:translatedTo, rdfs:comment, up:database (we inferred this variable might be of the type up:Transcript_Resource)\n",
      "Subject ?disease with type up:Disease in endpoint https://sparql.uniprot.org/sparql does not support the predicate rdfs:label. It can have the following predicates: skos:altLabel, rdfs:comment, up:mnemonic, skos:prefLabel, rdfs:seeAlso\n",
      "Subject ?orthologRatProtein with type orth:Protein in endpoint https://sparql.omabrowser.org/sparql/ does not support the predicate sio:SIO_010078. It can have the following predicates: lscr:xrefUniprot, orth:organism, obo:RO_0001018, lscr:xrefEnsemblProtein, lscr:xrefEnsemblTranscript, sio:SIO_010079, lscr:xrefSwissProt, rdfs:label, dc:identifier, rdfs:comment, lscr:xrefNCBIRefSeq, lscr:xrefNCBIProtein, rdfs:seeAlso, lscr:xrefEMBLSequence\n",
      "Subject ?pathVar5 in endpoint https://sparql.omabrowser.org/sparql/ does not support the predicate up:scientificNam. Correct predicate might be one of the following: up:rank, orth:taxRangeId, up:mnemonic, orth:taxRange, up:scientificName (we inferred this variable might be of the type orth:TaxonomicRange)\n"
     ]
    }
   ],
   "source": [
    "from sparql_llm.validate_sparql import validate_sparql_with_void\n",
    "\n",
    "# sparql_query = \"\"\"\n",
    "# PREFIX up: <http://purl.uniprot.org/core/>\n",
    "# PREFIX rh: <http://rdf.rhea-db.org/>\n",
    "# SELECT ?uniprotCount ?rhea ?accession ?equation\n",
    "# WHERE {\n",
    "#   SERVICE <https://sparql.uniprot.org/sparql> {\n",
    "#       SELECT ?rhea (count(?uniprot) as ?uniprotCount) {\n",
    "#           ?uniprot up:annotation ?rhea .\n",
    "#           ?uniprot up:toast ?toooast .\n",
    "#       }\n",
    "#       GROUP BY ?rhea\n",
    "#   }\n",
    "#   ?rhea rh:accession ?accession .\n",
    "#   ?rhea rh:equation ?equation .\n",
    "# }\n",
    "# \"\"\"\n",
    "\n",
    "# Hard query\n",
    "sparql_query = \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX up:<http://purl.uniprot.org/core/>\n",
    "PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth:<http://purl.org/net/orth#>\n",
    "PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "PREFIX lscr:<http://purl.org/lscr#>\n",
    "PREFIX genex:<http://purl.org/genex#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "SELECT DISTINCT ?diseaseLabel ?humanProtein ?hgncSymbol ?orthologRatProtein ?orthologRatGene\n",
    "WHERE {\n",
    "    SERVICE <https://sparql.uniprot.org/sparql> {\n",
    "        SELECT DISTINCT * WHERE {\n",
    "            ?humanProtein a up:Protein ;\n",
    "                up:organism/up:scientificName 'Homo sapiens' ;\n",
    "                up:annotation ?annotation ;\n",
    "                rdfs:seeAlso ?hgnc .\n",
    "            ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "                rdfs:label ?hgncSymbol . # comment\n",
    "            ?annotation a up:Disease_Annotation ;\n",
    "                up:disease ?disease .\n",
    "            ?disease a up:Disease ;\n",
    "                rdfs:label ?diseaseLabel . # skos:prefLabel\n",
    "            FILTER CONTAINS(?diseaseLabel, \"cancer\")\n",
    "        }\n",
    "    }\n",
    "    SERVICE <https://sparql.omabrowser.org/sparql/> {\n",
    "        SELECT ?humanProtein ?orthologRatProtein ?orthologRatGene WHERE {\n",
    "            ?humanProteinOma a orth:Protein ;\n",
    "                lscr:xrefUniprot ?humanProtein .\n",
    "            ?orthologRatProtein a orth:Protein ;\n",
    "                sio:SIO_010078 ?orthologRatGene ; # 79\n",
    "                orth:organism/obo:RO_0002162/up:scientificNam 'Rattus norvegicus' .\n",
    "            ?cluster a orth:OrthologsCluster .\n",
    "            ?cluster orth:hasHomologousMember ?node1 .\n",
    "            ?cluster orth:hasHomologousMember ?node2 .\n",
    "            ?node1 orth:hasHomologousMember* ?humanProteinOma .\n",
    "            ?node2 orth:hasHomologousMember* ?orthologRatProtein .\n",
    "            FILTER(?node1 != ?node2)\n",
    "        }\n",
    "    }\n",
    "    SERVICE <https://www.bgee.org/sparql/> {\n",
    "        ?orthologRatGene genex:isExpressedIn ?anatEntity ;\n",
    "            orth:organism ?ratOrganism .\n",
    "        ?anatEntity rdfs:label 'brain' .\n",
    "        ?ratOrganism obo:RO_0002162 taxon:10116 .\n",
    "    }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "issues = validate_sparql_with_void(sparql_query, \"https://sparql.uniprot.org/sparql\")\n",
    "print(\"\\n\".join(issues))\n",
    "\n",
    "# sparql_query = \"\"\"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "# PREFIX up: <http://purl.uniprot.org/core/>\n",
    "# PREFIX genex: <http://purl.org/genex#>\n",
    "# PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "# PREFIX orth: <http://purl.org/net/orth#>\n",
    "# PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "# PREFIX sio: <http://semanticscience.org/resource/>\n",
    "\n",
    "# SELECT DISTINCT ?flyEnsemblGene ?orthologTaxon ?orthologEnsemblGene ?orthologOmaLink WHERE {\n",
    "# \t{\n",
    "#         SELECT DISTINCT ?gene ?flyEnsemblGene {\n",
    "#         ?gene a orth:Gene ;\n",
    "#             genex:isExpressedIn/rdfs:label 'eye' ;\n",
    "#             orth:organism/obo:RO_0002162 ?taxon ;\n",
    "#             dcterms:identifier ?flyEnsemblGene .\n",
    "#         ?taxon up:commonName 'fruit fly' .\n",
    "#         } LIMIT 100\n",
    "#     }\n",
    "#     SERVICE <https://sparql.omabrowser.org/sparql> {\n",
    "#         ?protein2 a orth:Protein .\n",
    "#         ?protein1 a orth:Protein .\n",
    "#         ?clusterPrimates a orth:OrthologsCluster .\n",
    "#         ?cluster a orth:OrthologsCluster ;\n",
    "#             orth:hasHomologousMember ?node1 ;\n",
    "#             orth:hasHomologousMember ?node2 .\n",
    "#         ?node1 orth:hasHomologousMember* ?protein1 .\n",
    "#         ?node2 orth:hasHomologousMember* ?clusterPrimates .\n",
    "#         ?clusterPrimates orth:hasHomologousMember* ?protein2 .\n",
    "#         ?protein1 sio:SIO_010079 ?gene . # is encoded by\n",
    "#         ?protein2 rdfs:seeAlso ?orthologOmaLink ;\n",
    "#             orth:organism/obo:RO_0002162 ?orthologTaxonUri ;\n",
    "#             sio:SIO_010079 ?orthologGene . # is encoded by\n",
    "#         ?clusterPrimates orth:hasTaxonomicRange ?taxRange .\n",
    "#         ?taxRange orth:taxRange 'Primates' .\n",
    "#         FILTER ( ?node1 != ?node2 )\n",
    "#     }\n",
    "#     ?orthologTaxonUri up:commonName ?orthologTaxon .\n",
    "#     ?orthologGene dcterms:identifier ?orthologEnsemblGene .\n",
    "# }\"\"\"\n",
    "\n",
    "# try:\n",
    "#     validate_sparql_with_void(sparql_query, \"https://www.bgee.org/sparql\")\n",
    "# except Exception as e:\n",
    "#     print(e)\n",
    "\n",
    "\n",
    "# # Failing query:\n",
    "# sparql_query = \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "# PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "# PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "# SELECT ?protein ?sequence\n",
    "# WHERE {\n",
    "#     ?protein a up:Protein .\n",
    "#     ?protein up:reviewed true .\n",
    "#     ?protein up:organism taxon:10090 .\n",
    "#     ?protein up:matureProtein ?matureProtein .\n",
    "#     ?matureProtein rdf:value ?sequence .\n",
    "#     FILTER (STRSTARTS(?sequence, \"G\")) .\n",
    "# }\"\"\"\n",
    "\n",
    "\n",
    "# print(\"\\n\".join(error_msgs))\n",
    "# print(len(error_msgs))\n",
    "# qres = g.query(\"\"\"SELECT ?s ?p ?o WHERE {\n",
    "#     ?s ?p ?o\n",
    "# } LIMIT 10\"\"\")\n",
    "# for row in qres:\n",
    "#     print(f\"{row.s} knows {row.o}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"https://sparql.uniprot.org/sparql\": {\n",
      "    \"?anyKindOfName\": {\n",
      "      \"http://www.w3.org/2000/01/rdf-schema#subPropertyOf\": [\n",
      "        \"http://purl.uniprot.org/core/structuredNameType\"\n",
      "      ]\n",
      "    },\n",
      "    \"?protein\": {\n",
      "      \"http://www.w3.org/1999/02/22-rdf-syntax-ns#type\": [\n",
      "        \"http://purl.uniprot.org/core/Protein\"\n",
      "      ],\n",
      "      \"http://purl.uniprot.org/core/recommendedName\": [\n",
      "        \"?structuredName\"\n",
      "      ],\n",
      "      \"http://purl.uniprot.org/core/alternativeName\": [\n",
      "        \"?structuredName\"\n",
      "      ],\n",
      "      \"http://purl.uniprot.org/core/domain\": [\n",
      "        \"?pathVar2\"\n",
      "      ],\n",
      "      \"http://purl.uniprot.org/core/component\": [\n",
      "        \"?pathVar2\"\n",
      "      ]\n",
      "    },\n",
      "    \"?structuredName\": {\n",
      "      \"?anyKindOfName\": [\n",
      "        \"HLA class I histocompatibility antigen, B-73 alpha chain\"\n",
      "      ]\n",
      "    },\n",
      "    \"?pathVar2\": {\n",
      "      \"http://purl.uniprot.org/core/recommendedName\": [\n",
      "        \"?structuredName\"\n",
      "      ],\n",
      "      \"http://purl.uniprot.org/core/alternativeName\": [\n",
      "        \"?structuredName\"\n",
      "      ]\n",
      "    }\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "from sparql_llm.validate_sparql import sparql_query_to_dict\n",
    "\n",
    "path_query = \"\"\"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "SELECT ?protein ?anyKindOfName WHERE {\n",
    "\t\t?protein a up:Protein .\n",
    "\t\t?protein (up:recommendedName|up:alternativeName)|((up:domain|up:component)/(up:recommendedName|up:alternativeName)) ?structuredName .\n",
    "\t\t?structuredName ?anyKindOfName  \"HLA class I histocompatibility antigen, B-73 alpha chain\" .\n",
    "\t\t?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\n",
    "}\"\"\"\n",
    "\n",
    "sparql_query = \"\"\"\"\"\"\n",
    "\n",
    "# sparql_query\n",
    "query_dict = sparql_query_to_dict(path_query, \"https://sparql.uniprot.org/sparql\")\n",
    "# query_dict = sparql_query_to_dict(sparql_query, \"https://sparql.uniprot.org/sparql\")\n",
    "\n",
    "# print(query_dict)\n",
    "print(json.dumps(query_dict, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from rdflib.plugins.sparql import prepareQuery\n",
    "\n",
    "# NOTE: there is a bug in rdflib parser when nested SERVICE https://github.com/RDFLib/rdflib/issues/2136\n",
    "# query = \"\"\"\n",
    "# PREFIX wikibase: <http://wikiba.se/ontology#>\n",
    "# PREFIX wdt: <http://www.wikidata.org/prop/direct/>\n",
    "# PREFIX bd: <http://www.bigdata.com/rdf#>\n",
    "# PREFIX wd: <http://www.wikidata.org/entity/>\n",
    "# PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "# SELECT ?item ?pic\n",
    "# WHERE\n",
    "# {\n",
    "#     SERVICE <https://query.wikidata.org/sparql> {\n",
    "#         ?item wdt:P31 wd:Q146 .\n",
    "# \t\t?item wdt:P18 ?pic\n",
    "#         SERVICE wikibase:label {\n",
    "#             bd:serviceParam wikibase:language \"[AUTO_LANGUAGE],en\".\n",
    "#         }\n",
    "#     }\n",
    "# }\n",
    "\n",
    "# \"\"\"\n",
    "# parsed_query = parseQuery(query)\n",
    "# print(sq.algebra.name)\n",
    "# print(prepareQuery(toast_query).algebra)\n",
    "# triples = extract_triples(prepareQuery(sparql_query).algebra)\n",
    "\n",
    "# print(triples)\n",
    "# print(len(triples))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-chat",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
