Index: isubrip/namedtuples.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>from typing import List, NamedTuple, Type, Union\r\n\r\nfrom isubrip.enums import DataSource, SubtitlesType\r\n\r\n\r\nclass PlaylistData(NamedTuple):\r\n    \"\"\"A named tuple containing a playlist iTunes ID and URL.\"\"\"\r\n    itunes_id: str\r\n    url: str\r\n\r\n\r\nclass MovieData(NamedTuple):\r\n    \"\"\"A named tuple containing a movie name, id, and M3U8 playlist.\"\"\"\r\n    data_source: DataSource\r\n    name: str\r\n    release_year: int\r\n    playlists: List[PlaylistData]\r\n\r\n\r\nclass SubtitlesData(NamedTuple):\r\n    \"\"\"A named tuple containing language code, language name, type and playlist URL for subtitles.\"\"\"\r\n    language_code: str\r\n    language_name: str\r\n    subtitles_type: SubtitlesType\r\n    playlist_url: str\r\n\r\n\r\nclass ConfigSetting(NamedTuple):\r\n    category: str\r\n    key: str\r\n    types: Union[tuple, Type]\r\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/isubrip/namedtuples.py b/isubrip/namedtuples.py
--- a/isubrip/namedtuples.py	(revision 3120b7cdc58a946b9448f3712ed2c4ce663acb94)
+++ b/isubrip/namedtuples.py	(date 1662221832360)
@@ -11,10 +11,29 @@
 
 class MovieData(NamedTuple):
     """A named tuple containing a movie name, id, and M3U8 playlist."""
-    data_source: DataSource
+    name: str
+    release_year: int
+    playlists: Union[PlaylistData, List[PlaylistData]]
+
+
+class EpisodeData(NamedTuple):
+    episode_number: int
+    episode_name: Union[str, None]
+    playlist: Union[str, None]
+
+
+class SeasonData(NamedTuple):
+    """A named tuple containing a season number and M3U8 playlist."""
+    season_number: int
+    episode_name: Union[str, None]
+    episodes: List[EpisodeData]
+
+
+class SeriesData(NamedTuple):
+    """A named tuple containing a movie name, id, and M3U8 playlist."""
     name: str
     release_year: int
-    playlists: List[PlaylistData]
+    seasons: List[SeasonData]
 
 
 class SubtitlesData(NamedTuple):
