diff --git a/qlib/contrib/data/handler.py b/qlib/contrib/data/handler.py
index 2fe5258d..7ed86bb5 100644
--- a/qlib/contrib/data/handler.py
+++ b/qlib/contrib/data/handler.py
@@ -149,9 +149,9 @@ class Alpha158(DataHandlerLP):
         return Alpha158DL.get_feature_config(conf)
 
     def get_label_config(self):
-        return ["Ref($close, -2)/Ref($close, -1) - 1"], ["LABEL0"]
+        return ["Ref($close, -2)/(Ref($close, -1)+1e-12) - 1"], ["LABEL0"]
 
 
 class Alpha158vwap(Alpha158):
     def get_label_config(self):
-        return ["Ref($vwap, -2)/Ref($vwap, -1) - 1"], ["LABEL0"]
+        return ["Ref($vwap, -2)/(Ref($vwap, -1)+1e-12) - 1"], ["LABEL0"]
