Index: pyproject.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>[tool.ruff]\r\nline-length = 120\r\ntarget-version = \"py38\"\r\nselect = [\"B\", \"E\", \"F\"]\r\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pyproject.toml b/pyproject.toml
--- a/pyproject.toml	(revision 0e045d674a89b35a933f59ad2960d4a41d11bf85)
+++ b/pyproject.toml	(date 1685100643618)
@@ -1,4 +1,53 @@
 [tool.ruff]
 line-length = 120
 target-version = "py38"
-select = ["B", "E", "F"]
+select = [
+    "ARG",
+    "ASYNC",
+    "B",
+    "C4",
+    "COM",
+    "E",
+    "ERA",
+    "F",
+    "FA",
+    "I",
+    "INP",
+    "ISC",
+    "N",
+    "PIE",
+    "PGH",
+    "PL",
+    "PT",
+    "PTH",
+    "Q",
+    "RSE",
+    "RET",
+    "RUF",
+    "S",
+    "SIM",
+    "SLF",
+    "T20",
+    "TCH",
+    "TID",
+    "TRY",
+    "UP",
+]
+ignore = [
+    "C416",
+    "TD002",
+    "TD003",
+    "UP015",
+]
+unfixable = ["ARG"]
+
+[tool.ruff.isort]
+force-sort-within-sections = true
+
+[tool.ruff.flake8-quotes]
+docstring-quotes = "double"
+inline-quotes = "double"
+multiline-quotes = "double"
+
+[tool.ruff.flake8-tidy-imports]
+ban-relative-imports = "all"
\ No newline at end of file
