Index: pyproject.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>[tool.ruff]\r\nline-length = 120\r\ntarget-version = \"py38\"\r\nselect = [\"B\", \"E\", \"F\"]\r\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pyproject.toml b/pyproject.toml
--- a/pyproject.toml	(revision 0c70e3740d00de717748f4aec627850ddda0b0e9)
+++ b/pyproject.toml	(date 1685093480673)
@@ -1,4 +1,10 @@
 [tool.ruff]
 line-length = 120
 target-version = "py38"
-select = ["B", "E", "F"]
+select = ["B", "E", "F", "I"]
+
+[tool.ruff.isort]
+force-sort-within-sections = true
+
+[tool.ruff.flake8-tidy-imports]
+ban-relative-imports = "all"
\ No newline at end of file
